// @mtfe/sso-web-oidc-base/lib/config/const.js
"use strict";
var _a, _b;
Object.defineProperty(exports, "__esModule", { value: true });
exports.SSOUserActionType = exports.SSOIgnoreSLOType = exports.SSOPromptEnum = exports.allSSOSecondFactorTypes = exports.SSOSecondFactorEnum = exports.SSOLogLevel = exports.SSOLanguage = exports.SSOSameSiteType = exports.SSOWorkModel = exports.FIXED_SEARCH_PARAMS = exports.DEFAULT_LOGIN_PROTOCOL = exports.LOGIN_PROTOCOL_KEY = exports.SSOErrorCode = exports.CallbackErrorMsg = exports.CLI_DEFAULT_LOCAL_PORT = exports.AUTHORIZATION_ATTESTATION_HEADER_KEY = exports.AUTHORIZATION_REFRESH_HEADER_KEY = exports.AUTHORIZATION_HEADER_KEY = exports.SSO_ORIGINAL_URL_KEY = exports.SSO_WEB_OIDC_DEBUG_KEY = exports.LEGAL_ACCESS_ENV_LIST = exports.BUSINESS_SUCCESS_CODE = exports.SSO_CLI_MODE_ORIGIN_MAP = exports.SSOAccessEnvType = exports.REPORT_TYPE_CLI = exports.REPORT_TYPE = exports.SSOURLEnum = exports.OAUTH_PKCE_METHOD = exports.OAUTH_SCOPE = exports.OAUTH_RESPONSE_TYPE = exports.SSO_OIDC_LOGIN_PATH = exports.SSO_ORIGIN_MAP = void 0;
/**
 * 后缀不带/
 */
exports.SSO_ORIGIN_MAP = {
    test: 'https://ssosv.it.test.sankuai.com',
    product: 'https://ssosv.sankuai.com',
};
/**
 * SSO OIDC登录路径
 */
exports.SSO_OIDC_LOGIN_PATH = '/sson/auth/oidc/v1/authorize';
/**
 * OAuth responseType
 */
exports.OAUTH_RESPONSE_TYPE = 'code';
/**
 * OAuth scope
 */
exports.OAUTH_SCOPE = 'profile';
/**
 * OAuth PKCE code challenge method
 */
exports.OAUTH_PKCE_METHOD = 'S256';
var SSOURLEnum;
(function (SSOURLEnum) {
    /**
     * TOKEN刷新接口和验证接口
     */
    SSOURLEnum["TOKEN"] = "/websso/token/get";
    /**
     * OAuth回调地址
     */
    SSOURLEnum["CALLBACK"] = "/websso/oauth/callback";
    /**
     * 用户信息请求地址
     */
    SSOURLEnum["WHOAMI"] = "/websso/whoami";
    /**
     * 登出地址
     */
    SSOURLEnum["LOGOUT"] = "/websso/logout";
    /**
     * 上报地址
     */
    SSOURLEnum["REPORT"] = "/websso/report";
    /**
     * code换token地址
     */
    SSOURLEnum["EXCHANGE_TOKEN_BY_CODE"] = "/websso/oauth/pkce";
    /**
     * 交换token地址
     */
    SSOURLEnum["TOKEN_EXCHANGE"] = "/websso/token/exchange";
    /**
     * 交换token地址(cli版本)
     */
    SSOURLEnum["TOKEN_EXCHANGE_CLI"] = "/websso/auth/token-exchange";
    /**
     * 回调服务器状态获取地址
     */
    SSOURLEnum["CALLBACK_SERVER_WAITING"] = "/wait-for-auth";
    SSOURLEnum["SUCCESS_CALLBACK"] = "/websso/operation-callback";
})(SSOURLEnum || (exports.SSOURLEnum = SSOURLEnum = {}));
exports.REPORT_TYPE = 'web-sdk-oidc';
exports.REPORT_TYPE_CLI = 'web-sdk-oidc-cli';
/**
 * 访问环境
 */
var SSOAccessEnvType;
(function (SSOAccessEnvType) {
    SSOAccessEnvType["test"] = "test";
    SSOAccessEnvType["product"] = "product";
})(SSOAccessEnvType || (exports.SSOAccessEnvType = SSOAccessEnvType = {}));
/**
 * CLI模式下的SSO登录 origin
 */
exports.SSO_CLI_MODE_ORIGIN_MAP = {
    test: 'https://ssosv.it.test.sankuai.com',
    product: 'https://ssosv.sankuai.com',
};
/**
 * 后端接口请求成功code
 */
exports.BUSINESS_SUCCESS_CODE = 200;
/**
 * 合法的 accessEnv 值
 * */
exports.LEGAL_ACCESS_ENV_LIST = [SSOAccessEnvType.test, SSOAccessEnvType.product];
/**
 * 调试模式Session storage Key
 */
exports.SSO_WEB_OIDC_DEBUG_KEY = 'isDebug';
/**
 * original url key
 */
exports.SSO_ORIGINAL_URL_KEY = 'original-url';
/**
 * AccessToken header key
 */
exports.AUTHORIZATION_HEADER_KEY = 'Authorization';
/**
 * AccessToken refresh token header key
 */
exports.AUTHORIZATION_REFRESH_HEADER_KEY = 'Authorization-Refresh';
/**
 * Token exchange attestation header key
 */
exports.AUTHORIZATION_ATTESTATION_HEADER_KEY = 'Authorization-Attestation';
/**
 * Cli默认端口
 */
exports.CLI_DEFAULT_LOCAL_PORT = 9152;
/**
 * OAuth回调错误信息
 */
var CallbackErrorMsg;
(function (CallbackErrorMsg) {
    CallbackErrorMsg["MISS_CODE"] = "Error: Missing required code parameter!";
    CallbackErrorMsg["MISS_CODE_VERIFIER"] = "Error: Missing required code_verifier, maybe a csrf attack!";
    CallbackErrorMsg["STATE_NO_MATCH"] = "Error: state not match, maybe a csrf attack!";
})(CallbackErrorMsg || (exports.CallbackErrorMsg = CallbackErrorMsg = {}));
/**
 * SSO操作状态码
 */
var SSOErrorCode;
(function (SSOErrorCode) {
    /**
     * 操作成功
     */
    SSOErrorCode[SSOErrorCode["SUCCESS"] = 0] = "SUCCESS";
    /**
     * 未登录/未授权
     */
    SSOErrorCode[SSOErrorCode["UNAUTHORIZED"] = 401] = "UNAUTHORIZED";
    /**
     * 请求超时
     */
    SSOErrorCode[SSOErrorCode["TIMEOUT"] = 1001] = "TIMEOUT";
    /**
     * 通用错误
     */
    SSOErrorCode[SSOErrorCode["UNKNOWN_ERROR"] = 1002] = "UNKNOWN_ERROR";
    /**
     * 获取用户信息失败
     */
    SSOErrorCode[SSOErrorCode["GET_USERINFO_ERROR"] = 1003] = "GET_USERINFO_ERROR";
    /**
     * code参数缺失
     */
    SSOErrorCode[SSOErrorCode["MISS_CODE"] = 1004] = "MISS_CODE";
    /**
     * code_verifier缺失
     */
    SSOErrorCode[SSOErrorCode["MISS_CODE_VERIFIER"] = 1005] = "MISS_CODE_VERIFIER";
    /**
     * state参数丢失或不一致
     */
    SSOErrorCode[SSOErrorCode["STATE_NO_MATCH"] = 1006] = "STATE_NO_MATCH";
    /**
     * 参数缺失
     */
    SSOErrorCode[SSOErrorCode["MISS_PARAMS"] = 1007] = "MISS_PARAMS";
    /**
     * 创建监听服务器失败
     */
    SSOErrorCode[SSOErrorCode["CREATE_SERVER_ERROR"] = 1008] = "CREATE_SERVER_ERROR";
    /**
     * 无可用端口
     */
    SSOErrorCode[SSOErrorCode["NO_AVALIABLE_PORT"] = 1009] = "NO_AVALIABLE_PORT";
    /**
     * 重试次数过多
     */
    SSOErrorCode[SSOErrorCode["RETRY_TOO_MANY_TIME"] = 1010] = "RETRY_TOO_MANY_TIME";
    /**
     * 正在处理中
     */
    SSOErrorCode[SSOErrorCode["PROCESSING"] = 1011] = "PROCESSING";
    /**
     * 手动取消
     */
    SSOErrorCode[SSOErrorCode["MANUAL_CANCEL"] = 1012] = "MANUAL_CANCEL";
    /**
     * Token Exchange失败
     */
    SSOErrorCode[SSOErrorCode["TOKEN_EXCHANGE_FAIL"] = 80002] = "TOKEN_EXCHANGE_FAIL";
    /**
     * 用户身份访问许可（即用户一级授权）
     */
    SSOErrorCode[SSOErrorCode["USER_ACCESS_PERMISSION"] = 80003] = "USER_ACCESS_PERMISSION";
    /**
     * 应用身份访问许可（应用身份权限）
     */
    SSOErrorCode[SSOErrorCode["APPLICATION_ACCESS_PERMISSION"] = 80004] = "APPLICATION_ACCESS_PERMISSION";
    /**
     * 用户和发起换票的应用没有目标应用的权限，相当于 80003 | 8004
     */
    SSOErrorCode[SSOErrorCode["USER_AND_APPLICATION_ACCESS_PERMISSION"] = 80005] = "USER_AND_APPLICATION_ACCESS_PERMISSION";
})(SSOErrorCode || (exports.SSOErrorCode = SSOErrorCode = {}));
/**
 * 登录协议参数名
 */
exports.LOGIN_PROTOCOL_KEY = 'loginProtocol';
/**
 * 默认登录协议
 */
exports.DEFAULT_LOGIN_PROTOCOL = 'OIDC';
/**
 * 固定默认参数
 */
exports.FIXED_SEARCH_PARAMS = (_a = {},
    _a[exports.LOGIN_PROTOCOL_KEY] = exports.DEFAULT_LOGIN_PROTOCOL,
    _a);
/**
 * SSO工作模式，可选值分别是浏览器、cli，未设置默认是浏览器
 */
var SSOWorkModel;
(function (SSOWorkModel) {
    SSOWorkModel["browser"] = "browser";
    SSOWorkModel["cli"] = "cli";
})(SSOWorkModel || (exports.SSOWorkModel = SSOWorkModel = {}));
/**
 * SameSite属性
 * 支持lax、strict、none、不设置
 */
var SSOSameSiteType;
(function (SSOSameSiteType) {
    SSOSameSiteType["lax"] = "lax";
    SSOSameSiteType["strict"] = "strict";
    SSOSameSiteType["none"] = "none";
})(SSOSameSiteType || (exports.SSOSameSiteType = SSOSameSiteType = {}));
/**
 * 支持的语言
 */
var SSOLanguage;
(function (SSOLanguage) {
    SSOLanguage["zh"] = "zh";
    SSOLanguage["en"] = "en";
    SSOLanguage["zh_HK"] = "zh-HK";
    SSOLanguage["pt-BR"] = "pt-BR";
})(SSOLanguage || (exports.SSOLanguage = SSOLanguage = {}));
/**
 * 日志等级的使用原则
 * 数值越大，级别越高，输出越多
 * 生产环境：通常设置为 INFO 或 WARN 级别
 * 开发环境：通常设置为 DEBUG 级别
 */
var SSOLogLevel;
(function (SSOLogLevel) {
    SSOLogLevel[SSOLogLevel["SILENT"] = 0] = "SILENT";
    SSOLogLevel[SSOLogLevel["ERROR"] = 1] = "ERROR";
    SSOLogLevel[SSOLogLevel["INFO"] = 2] = "INFO";
    SSOLogLevel[SSOLogLevel["WARN"] = 3] = "WARN";
    SSOLogLevel[SSOLogLevel["DEBUG"] = 4] = "DEBUG";
})(SSOLogLevel || (exports.SSOLogLevel = SSOLogLevel = {}));
/**
 * 二次认证类型
 */
var SSOSecondFactorEnum;
(function (SSOSecondFactorEnum) {
    /**
     * sms短信认证
     */
    SSOSecondFactorEnum["SMS"] = "sms";
    /**
     * voice语音认证
     */
    SSOSecondFactorEnum["VOICE"] = "voice";
    /**
     * otp动态码认证
     */
    SSOSecondFactorEnum["OTP"] = "otp";
    /**
     * email邮件认证
     */
    SSOSecondFactorEnum["EMAIL"] = "email";
    /**
     * assistance同事辅助认证
     */
    SSOSecondFactorEnum["ASSISTANCE"] = "assistance";
    /**
     * face人脸认证
     */
    SSOSecondFactorEnum["FACE"] = "face";
    /**
     * faceqrcode人脸扫码认证
     */
    SSOSecondFactorEnum["FACE_QRCODE"] = "faceqrcode";
    /**
     * qrcode大象扫码认证
     */
    SSOSecondFactorEnum["QRCODE"] = "qrcode";
    /**
     * Passkey
     */
    SSOSecondFactorEnum["PASSKEY"] = "passkey";
})(SSOSecondFactorEnum || (exports.SSOSecondFactorEnum = SSOSecondFactorEnum = {}));
// 如果枚举有新增项但这里没有包含，TypeScript 会报错
exports.allSSOSecondFactorTypes = (_b = {},
    _b[SSOSecondFactorEnum.SMS] = SSOSecondFactorEnum.SMS,
    _b[SSOSecondFactorEnum.VOICE] = SSOSecondFactorEnum.VOICE,
    _b[SSOSecondFactorEnum.OTP] = SSOSecondFactorEnum.OTP,
    _b[SSOSecondFactorEnum.EMAIL] = SSOSecondFactorEnum.EMAIL,
    _b[SSOSecondFactorEnum.ASSISTANCE] = SSOSecondFactorEnum.ASSISTANCE,
    _b[SSOSecondFactorEnum.FACE] = SSOSecondFactorEnum.FACE,
    _b[SSOSecondFactorEnum.FACE_QRCODE] = SSOSecondFactorEnum.FACE_QRCODE,
    _b[SSOSecondFactorEnum.QRCODE] = SSOSecondFactorEnum.QRCODE,
    _b[SSOSecondFactorEnum.PASSKEY] = SSOSecondFactorEnum.PASSKEY,
    _b);
// OIDC prompt类型
var SSOPromptEnum;
(function (SSOPromptEnum) {
    /**
     * Consent
     */
    SSOPromptEnum["CONSENT"] = "consent";
})(SSOPromptEnum || (exports.SSOPromptEnum = SSOPromptEnum = {}));
/**
 * 忽略单点登出类型
 * 场景说明：用户获取应用B的小票时，指定ignoreSLOType = xxx、与应用A共用同一张根票，如果用户退出应用A
 */
var SSOIgnoreSLOType;
(function (SSOIgnoreSLOType) {
    /**
     * 表示不忽略，即用户访问应用B的小票同时失效；
     */
    SSOIgnoreSLOType["none"] = "none";
    /**
     * 表示部分忽略，即用户访问应用B的小票在introspect、refresh场景下仍有效。
     */
    SSOIgnoreSLOType["part"] = "part";
})(SSOIgnoreSLOType || (exports.SSOIgnoreSLOType = SSOIgnoreSLOType = {}));
/**
 * 用户行为类型
 */
var SSOUserActionType;
(function (SSOUserActionType) {
    SSOUserActionType["login"] = "login";
    SSOUserActionType["loginWithModal"] = "login_modal";
    SSOUserActionType["tokenExchange"] = "token_exchange";
})(SSOUserActionType || (exports.SSOUserActionType = SSOUserActionType = {}));
