// @mtfe/sso-web-oidc-base/lib/utils/httpClient.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g = Object.create((typeof Iterator === "function" ? Iterator : Object).prototype);
    return g.next = verb(0), g["throw"] = verb(1), g["return"] = verb(2), typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var const_1 = require("../config/const");
var query_string_1 = __importDefault(require("query-string"));
var utils_1 = require("./utils");
var version_1 = require("../version");
var env_1 = require("./env");
var typeUtil_1 = require("./typeUtil");
/**
 * fetch请求方法
 */
var RequestMethodEnum;
(function (RequestMethodEnum) {
    RequestMethodEnum["POST"] = "POST";
    RequestMethodEnum["GET"] = "GET";
})(RequestMethodEnum || (RequestMethodEnum = {}));
function timeout(originalFetch, ms) {
    var timer;
    var highOrderFetch = Promise.race([
        originalFetch,
        new Promise(function (_resolve, reject) {
            timer = setTimeout(function () {
                reject((0, utils_1.buildResult)(const_1.SSOErrorCode.TIMEOUT, null, 'timeout'));
            }, ms || 15 * 1000);
        }),
    ]);
    highOrderFetch.finally(function () {
        if (timer) {
            clearTimeout(timer);
        }
    });
    return highOrderFetch;
}
/**
 * HTTP客户端
 */
var HttpClient = /** @class */ (function () {
    function HttpClient(_a) {
        var configs = _a.configs, request = _a.request, logger = _a.logger, sdkVersion = _a.sdkVersion;
        this.configs = __assign({}, configs);
        this.sdkVersion = sdkVersion;
        var mode = configs.mode;
        this.request = request;
        this.logger = logger;
        var browserOrigin = env_1.ENV.isBrowser ? location.origin : '';
        this.origin =
            mode === const_1.SSOWorkModel.cli
                ? (0, utils_1.getRealOrigin)(const_1.SSO_CLI_MODE_ORIGIN_MAP, this.configs)
                : browserOrigin;
    }
    /**
     * 获取AccessToken
     * @param accessToken accessToken
     * @returns AccessToken等信息
     */
    HttpClient.prototype.getToken = function (tokenOption) {
        return __awaiter(this, void 0, void 0, function () {
            var origin, _a, clientId, sameSite, searchParams, url, headers, _b, access_token, refresh_token, token_exchange_attestation, TAG, data, err_1;
            return __generator(this, function (_c) {
                switch (_c.label) {
                    case 0:
                        origin = this.origin;
                        _a = this.configs, clientId = _a.clientId, sameSite = _a.sameSite;
                        searchParams = __assign(__assign({}, (0, utils_1.getFixedSearchParams)(__assign({ sdkVersion: this.sdkVersion }, this.configs))), { client_id: clientId });
                        if (sameSite) {
                            // 历史原因，接口参数继续使用samesite，非sameSite
                            searchParams.samesite = sameSite;
                        }
                        url = "".concat(origin).concat(const_1.SSOURLEnum.TOKEN, "?").concat(query_string_1.default.stringify(searchParams));
                        headers = {};
                        _b = tokenOption || {}, access_token = _b.access_token, refresh_token = _b.refresh_token, token_exchange_attestation = _b.token_exchange_attestation;
                        if (access_token) {
                            headers[const_1.AUTHORIZATION_HEADER_KEY] = (0, utils_1.createAuthorizationHeader)(access_token);
                        }
                        if (refresh_token) {
                            headers[const_1.AUTHORIZATION_REFRESH_HEADER_KEY] = (0, utils_1.createAuthorizationHeader)(refresh_token);
                        }
                        if (token_exchange_attestation) {
                            headers[const_1.AUTHORIZATION_ATTESTATION_HEADER_KEY] = (0, utils_1.createAuthorizationHeader)(token_exchange_attestation);
                        }
                        TAG = 'getToken';
                        _c.label = 1;
                    case 1:
                        _c.trys.push([1, 3, , 4]);
                        return [4 /*yield*/, this.fetchRequest({ url: url, method: RequestMethodEnum.POST, headers: headers })];
                    case 2:
                        data = _c.sent();
                        return [2 /*return*/, this.catchRequestResult(data, TAG)];
                    case 3:
                        err_1 = _c.sent();
                        return [2 /*return*/, this.catchRequestError(err_1, TAG)];
                    case 4: return [2 /*return*/];
                }
            });
        });
    };
    /**
     * 获取用户信息
     * @param accessToken accessToken
     * @returns 用户信息
     */
    HttpClient.prototype.getUserInfo = function (accessToken) {
        return __awaiter(this, void 0, void 0, function () {
            var origin, clientId, searchParams, url, headers, TAG, data, err_2;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        origin = this.origin;
                        clientId = this.configs.clientId;
                        searchParams = __assign(__assign({}, (0, utils_1.getFixedSearchParams)(__assign({ sdkVersion: this.sdkVersion }, this.configs))), { client_id: clientId });
                        url = "".concat(origin).concat(const_1.SSOURLEnum.WHOAMI, "?").concat(query_string_1.default.stringify(searchParams));
                        headers = {};
                        if (accessToken) {
                            headers[const_1.AUTHORIZATION_HEADER_KEY] = (0, utils_1.createAuthorizationHeader)(accessToken);
                        }
                        TAG = 'getUserInfo';
                        _a.label = 1;
                    case 1:
                        _a.trys.push([1, 3, , 4]);
                        return [4 /*yield*/, this.fetchRequest({ url: url, method: RequestMethodEnum.GET, headers: headers })];
                    case 2:
                        data = _a.sent();
                        return [2 /*return*/, this.catchRequestResult(data, TAG)];
                    case 3:
                        err_2 = _a.sent();
                        return [2 /*return*/, this.catchRequestError(err_2, TAG)];
                    case 4: return [2 /*return*/];
                }
            });
        });
    };
    /**
     * 登出（仅限CLI工作模式）
     * @param accessToken accessToken
     */
    HttpClient.prototype.logout = function (accessToken) {
        return __awaiter(this, void 0, void 0, function () {
            var clientId, searchParams, url, headers, TAG, data, err_3;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        clientId = this.configs.clientId;
                        searchParams = __assign(__assign({}, (0, utils_1.getFixedSearchParams)(__assign({ sdkVersion: this.sdkVersion }, this.configs))), { client_id: clientId, redirect_uri: (0, utils_1.getOriginalUrl)(this.configs) });
                        url = "".concat(this.getLogoutUrl(), "?").concat(query_string_1.default.stringify(searchParams));
                        headers = {};
                        if (accessToken) {
                            headers[const_1.AUTHORIZATION_HEADER_KEY] = (0, utils_1.createAuthorizationHeader)(accessToken);
                        }
                        TAG = 'logout';
                        _a.label = 1;
                    case 1:
                        _a.trys.push([1, 3, , 4]);
                        return [4 /*yield*/, this.fetchRequest({ url: url, method: RequestMethodEnum.GET, headers: headers })];
                    case 2:
                        data = _a.sent();
                        return [2 /*return*/, this.catchRequestResult(data, TAG)];
                    case 3:
                        err_3 = _a.sent();
                        return [2 /*return*/, this.catchRequestError(err_3, TAG)];
                    case 4: return [2 /*return*/];
                }
            });
        });
    };
    /**
     * 上报配置信息
     * @param configs SSO配置
     * @returns
     */
    HttpClient.prototype.reportConfig = function () {
        return __awaiter(this, void 0, void 0, function () {
            var _a, clientId, accessEnv, mode, url, extraConfig, postData, err_4;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        _a = this.configs, clientId = _a.clientId, accessEnv = _a.accessEnv, mode = _a.mode;
                        url = this.getReportURL();
                        extraConfig = __assign(__assign({}, (0, utils_1.getFixedSearchParams)(__assign({ sdkVersion: this.sdkVersion }, this.configs))), this.configs);
                        postData = {
                            clientId: clientId,
                            env: accessEnv,
                            version: (0, version_1.getPkgVersion)(),
                            config: JSON.stringify(extraConfig),
                            type: mode === const_1.SSOWorkModel.cli ? const_1.REPORT_TYPE_CLI : const_1.REPORT_TYPE,
                        };
                        _b.label = 1;
                    case 1:
                        _b.trys.push([1, 3, , 4]);
                        return [4 /*yield*/, this.fetchRequest({ url: url, method: RequestMethodEnum.POST, data: postData })];
                    case 2:
                        _b.sent();
                        return [3 /*break*/, 4];
                    case 3:
                        err_4 = _b.sent();
                        this.logger.warn('reportConfig', err_4);
                        return [3 /*break*/, 4];
                    case 4: return [2 /*return*/];
                }
            });
        });
    };
    /**
     * 某一具体信息
     * @param type 某一具体类型，如登录、tokenExchange
     */
    HttpClient.prototype.reportType = function (actionType) {
        return __awaiter(this, void 0, void 0, function () {
            var _a, clientId, accessEnv, mode, url, extraConfig, realActionType, postData, err_5;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        if (!actionType)
                            return [2 /*return*/];
                        _a = this.configs, clientId = _a.clientId, accessEnv = _a.accessEnv, mode = _a.mode;
                        url = this.getReportURL();
                        extraConfig = __assign(__assign({}, (0, utils_1.getFixedSearchParams)(__assign({ sdkVersion: this.sdkVersion }, this.configs))), this.configs);
                        realActionType = "".concat(mode === const_1.SSOWorkModel.cli ? const_1.REPORT_TYPE_CLI : const_1.REPORT_TYPE, "_").concat(actionType);
                        postData = {
                            clientId: clientId,
                            env: accessEnv,
                            version: (0, version_1.getPkgVersion)(),
                            config: JSON.stringify(extraConfig),
                            type: realActionType,
                        };
                        _b.label = 1;
                    case 1:
                        _b.trys.push([1, 3, , 4]);
                        return [4 /*yield*/, this.fetchRequest({ url: url, method: RequestMethodEnum.POST, data: postData })];
                    case 2:
                        _b.sent();
                        return [3 /*break*/, 4];
                    case 3:
                        err_5 = _b.sent();
                        return [3 /*break*/, 4];
                    case 4: return [2 /*return*/];
                }
            });
        });
    };
    /**
     * 通过code换token
     * @param options 请求配置
     * @returns 操作结果
     */
    HttpClient.prototype.exchangeTokenByCode = function (options) {
        return __awaiter(this, void 0, void 0, function () {
            var origin, _a, clientId, mode, searchParams, url, TAG, data, err_6;
            var _b;
            return __generator(this, function (_c) {
                switch (_c.label) {
                    case 0:
                        origin = this.origin;
                        _a = this.configs, clientId = _a.clientId, mode = _a.mode;
                        searchParams = __assign(__assign({}, (0, utils_1.getFixedSearchParams)(__assign({ sdkVersion: this.sdkVersion }, this.configs))), (_b = { code: options.code, code_verifier: options.codeVerifier, state: options.state, client_id: clientId, mode: mode }, _b[const_1.SSO_ORIGINAL_URL_KEY] = (0, utils_1.getOriginalUrl)(this.configs), _b));
                        url = "".concat(origin).concat(const_1.SSOURLEnum.EXCHANGE_TOKEN_BY_CODE, "?").concat(query_string_1.default.stringify(searchParams));
                        TAG = 'exchangeTokenByCode';
                        this.logger.debug(TAG, url);
                        _c.label = 1;
                    case 1:
                        _c.trys.push([1, 3, , 4]);
                        return [4 /*yield*/, this.fetchRequest({ url: url, method: RequestMethodEnum.GET })];
                    case 2:
                        data = _c.sent();
                        return [2 /*return*/, this.catchRequestResult(data, TAG)];
                    case 3:
                        err_6 = _c.sent();
                        return [2 /*return*/, this.catchRequestError(err_6, TAG)];
                    case 4: return [2 /*return*/];
                }
            });
        });
    };
    /**
     * 获取目标clientId的access_token
     * @param options Token Exchange参数
     * @returns 换成后的access_token
     */
    HttpClient.prototype.tokenExchange = function (options) {
        return __awaiter(this, void 0, void 0, function () {
            var origin, postData, mode, searchParams, realPath, url, TAG, data, err_7;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        origin = this.origin;
                        postData = {
                            client_id: options.clientId,
                            audience: options.audience,
                        };
                        if (options.access_token) {
                            postData.access_token = options.access_token;
                        }
                        if (options.token_exchange_attestation) {
                            postData.token_exchange_attestation = options.token_exchange_attestation;
                        }
                        mode = this.configs.mode;
                        searchParams = __assign({}, (0, utils_1.getFixedSearchParams)(__assign({ sdkVersion: this.sdkVersion }, this.configs)));
                        realPath = mode === const_1.SSOWorkModel.cli ? const_1.SSOURLEnum.TOKEN_EXCHANGE_CLI : const_1.SSOURLEnum.TOKEN_EXCHANGE;
                        url = "".concat(origin).concat(realPath, "?").concat(query_string_1.default.stringify(searchParams));
                        TAG = 'tokenExchange';
                        this.logger.debug(TAG, url);
                        _a.label = 1;
                    case 1:
                        _a.trys.push([1, 3, , 4]);
                        return [4 /*yield*/, this.fetchRequest({ url: url, data: postData, method: RequestMethodEnum.POST })];
                    case 2:
                        data = _a.sent();
                        return [2 /*return*/, this.catchRequestResult(data, TAG)];
                    case 3:
                        err_7 = _a.sent();
                        return [2 /*return*/, this.catchRequestError(err_7, TAG)];
                    case 4: return [2 /*return*/];
                }
            });
        });
    };
    /**
     * 封装后的请求方法
     * @param param0 请求参数
     * @returns
     */
    HttpClient.prototype.fetchRequest = function (_a) {
        var _this = this;
        var url = _a.url, method = _a.method, data = _a.data, headers = _a.headers;
        var fetchOptions = {
            method: method,
            headers: __assign({ 'Content-Type': 'application/json' }, (headers || {})),
        };
        if (data) {
            fetchOptions.body = JSON.stringify(data);
        }
        var reportUrl = this.getReportURL();
        var logoutUrl = this.getLogoutUrl();
        this.logger.debug('fetchRequest', method, url);
        var currentGlobal = (0, env_1.getGlobal)();
        var realFetch = currentGlobal.fetch ? currentGlobal.fetch : this.request;
        if (!realFetch)
            return Promise.reject({
                code: const_1.SSOErrorCode.UNKNOWN_ERROR,
                msg: "Can't find fetch function",
            });
        return timeout(realFetch(url, fetchOptions)).then(function (response) {
            var _a = response || {}, ok = _a.ok, status = _a.status, statusText = _a.statusText;
            _this.logger.debug('response', url, ' with:', status);
            if (ok) {
                if (url.startsWith(reportUrl) || url.startsWith(logoutUrl)) {
                    return { code: const_1.SSOErrorCode.SUCCESS };
                }
                return response.json();
            }
            var returnCode = status === 401 ? const_1.SSOErrorCode.UNAUTHORIZED : const_1.SSOErrorCode.UNKNOWN_ERROR;
            return Promise.reject({
                code: returnCode,
                msg: "".concat(status ? status : '', " ").concat(statusText ? statusText : ''),
                status: status,
            });
        });
    };
    HttpClient.prototype.getReportURL = function () {
        var origin = this.origin;
        return "".concat(origin).concat(const_1.SSOURLEnum.REPORT);
    };
    HttpClient.prototype.getLogoutUrl = function () {
        var origin = this.origin;
        return "".concat(origin).concat(const_1.SSOURLEnum.LOGOUT);
    };
    /**
     * 处理请求结果
     * @param data 请求返回数据
     * @param tag 错误TAG
     * @returns 返回结果
     */
    HttpClient.prototype.catchRequestResult = function (data, tag) {
        if (data && (data.code === const_1.BUSINESS_SUCCESS_CODE || data.code === const_1.SSOErrorCode.SUCCESS)) {
            var result = (0, utils_1.buildResult)(const_1.SSOErrorCode.SUCCESS, data.data, data.msg);
            this.logger.debug(tag, result === null || result === void 0 ? void 0 : result.code);
            return result;
        }
        // 后端返回异常数据
        throw data;
    };
    /**
     * 捕获并返回结果
     * @param err 错误信息
     * @param tag 错误TAG
     * @returns
     */
    HttpClient.prototype.catchRequestError = function (err, tag) {
        var code = (0, typeUtil_1.tsSafeCheckKey)(err, 'code')
            ? err.code
            : const_1.SSOErrorCode.UNKNOWN_ERROR;
        var msg = "".concat(tag, " fail");
        if ((0, typeUtil_1.tsSafeCheckKey)(err, 'msg')) {
            // eslint-disable-next-line prefer-destructuring
            msg = err.msg;
        }
        if (!msg && (0, typeUtil_1.tsSafeCheckKey)(err, 'message')) {
            msg = err.message;
        }
        this.logger.warn(tag, err);
        return (0, utils_1.buildResult)(code, null, msg);
    };
    return HttpClient;
}());
exports.default = HttpClient;
