// @mtfe/sso-web-oidc-base/lib/utils/logger.js
"use strict";
var __spreadArray = (this && this.__spreadArray) || function (to, from, pack) {
    if (pack || arguments.length === 2) for (var i = 0, l = from.length, ar; i < l; i++) {
        if (ar || !(i in from)) {
            if (!ar) ar = Array.prototype.slice.call(from, 0, i);
            ar[i] = from[i];
        }
    }
    return to.concat(ar || Array.prototype.slice.call(from));
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ConsoleOutput = void 0;
var const_1 = require("../config/const");
var env_1 = require("./env");
// 敏感信息关键词配置
var SENSITIVE_KEYWORDS = [
    'password',
    'token',
    'secret',
    'key',
    'authorization',
    'auth',
    'credential',
    'access_token',
    'refresh_token',
    'id_token',
    'client_secret',
    'private_key',
    'session',
    'cookie',
    'jwt',
    'bearer',
    'code',
    'Authorization-Refresh',
    'Authorization',
];
var DEFAULT_MASK_CONFIG = {
    prefixLength: 3,
    suffixLength: 3,
    maskChar: '*',
    maskLength: 6, // 固定输出6个星号
    minLength: 8,
};
/**
 * 检查字段名是否包含敏感关键词
 */
var isSensitiveField = function (fieldName) {
    if (!fieldName || typeof fieldName !== 'string') {
        return false;
    }
    var lowerFieldName = fieldName.toLowerCase();
    return SENSITIVE_KEYWORDS.some(function (keyword) { return lowerFieldName.includes(keyword.toLowerCase()); });
};
/**
 * 对字符串进行掩码处理
 */
var maskString = function (value, config) {
    if (config === void 0) { config = DEFAULT_MASK_CONFIG; }
    // 基础验证
    if (!value || typeof value !== 'string') {
        return value;
    }
    // 长度不足最小要求，不进行掩码
    if (value.length < config.minLength) {
        return value;
    }
    var prefixLength = config.prefixLength, suffixLength = config.suffixLength, maskChar = config.maskChar, maskLength = config.maskLength;
    var totalVisibleLength = prefixLength + suffixLength;
    // 边界场景1: 前后缀长度之和大于等于字符串长度
    if (totalVisibleLength >= value.length) {
        // 优先保证前缀，如果还有剩余长度再显示后缀
        var actualPrefixLength = Math.min(prefixLength, value.length);
        var remainingLength = value.length - actualPrefixLength;
        var actualSuffixLength = Math.min(suffixLength, remainingLength);
        if (actualSuffixLength === 0) {
            // 只能显示前缀
            return value.substring(0, actualPrefixLength);
        }
        // 显示前缀和后缀，中间不加掩码
        var prefix_1 = value.substring(0, actualPrefixLength);
        var suffix_1 = value.substring(value.length - actualSuffixLength);
        return "".concat(prefix_1).concat(suffix_1);
    }
    // 边界场景2: 前后缀长度之和加上固定掩码长度大于字符串长度
    if (totalVisibleLength + maskLength > value.length) {
        // 动态调整前后缀长度，确保能容纳固定长度的掩码
        var availableLength = value.length - maskLength;
        if (availableLength <= 0) {
            // 字符串太短，无法容纳固定掩码，只显示固定掩码
            return maskChar.repeat(Math.min(maskLength, value.length));
        }
        // 按比例分配前后缀长度
        var ratio = prefixLength / (prefixLength + suffixLength);
        var adjustedPrefixLength = Math.floor(availableLength * ratio);
        var adjustedSuffixLength = availableLength - adjustedPrefixLength;
        var prefix_2 = value.substring(0, adjustedPrefixLength);
        var suffix_2 = value.substring(value.length - adjustedSuffixLength);
        return "".concat(prefix_2).concat(maskChar.repeat(maskLength)).concat(suffix_2);
    }
    // 正常场景: 使用固定长度掩码
    var prefix = value.substring(0, prefixLength);
    var suffix = value.substring(value.length - suffixLength);
    return "".concat(prefix).concat(maskChar.repeat(maskLength)).concat(suffix);
};
/**
 * 深度处理对象，对敏感字段进行掩码
 */
var maskSensitiveData = function (data, maxDepth, currentDepth) {
    if (maxDepth === void 0) { maxDepth = 10; }
    if (currentDepth === void 0) { currentDepth = 0; }
    // 防止无限递归
    if (currentDepth >= maxDepth) {
        return '[Max Depth Reached]';
    }
    // 处理 null 和 undefined
    if (data === null || data === undefined) {
        return data;
    }
    // 处理基本类型
    if (typeof data !== 'object') {
        return data;
    }
    // 处理数组
    if (Array.isArray(data)) {
        return data.map(function (item) { return maskSensitiveData(item, maxDepth, currentDepth + 1); });
    }
    // 处理 Date 对象
    if (data instanceof Date) {
        return data;
    }
    // 处理 Error 对象
    if (data instanceof Error) {
        return {
            name: data.name,
            message: data.message,
            stack: data.stack,
        };
    }
    // 处理普通对象
    try {
        var maskedObj = {};
        for (var _i = 0, _a = Object.entries(data); _i < _a.length; _i++) {
            var _b = _a[_i], key = _b[0], value = _b[1];
            if (isSensitiveField(key) && typeof value === 'string') {
                maskedObj[key] = maskString(value);
            }
            else if (typeof value === 'object' && value !== null) {
                maskedObj[key] = maskSensitiveData(value, maxDepth, currentDepth + 1);
            }
            else {
                maskedObj[key] = value;
            }
        }
        return maskedObj;
    }
    catch (error) {
        // 如果处理过程中出现错误，返回安全的表示
        return '[Object Processing Error]';
    }
};
/**
 * 处理日志参数，对敏感信息进行掩码
 */
var processLogArgs = function (args, logLevel) {
    if (logLevel === const_1.SSOLogLevel.DEBUG) {
        return args;
    }
    return args.map(function (arg) {
        try {
            return maskSensitiveData(arg);
        }
        catch (error) {
            // 如果处理失败，返回原始参数的安全表示
            return typeof arg === 'object' ? '[Object]' : arg;
        }
    });
};
// 控制台输出器
var ConsoleOutput = /** @class */ (function () {
    function ConsoleOutput(mode) {
        this.mode = mode;
    }
    ConsoleOutput.prototype.log = function () {
        var args = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            args[_i] = arguments[_i];
        }
        if (this.mode === const_1.SSOWorkModel.cli) {
            // eslint-disable-next-line no-console
            console.error.apply(console, args);
        }
        else {
            // eslint-disable-next-line no-console
            console.log.apply(console, args);
        }
    };
    return ConsoleOutput;
}());
exports.ConsoleOutput = ConsoleOutput;
var SSOLogger = /** @class */ (function () {
    function SSOLogger(options) {
        var _a;
        this.packageName = options === null || options === void 0 ? void 0 : options.packageName;
        this.mode = options.mode;
        // 使用自定义输出器，如果没有设置则使用默认的控制台输出
        this.output = options.output || new ConsoleOutput(this.mode);
        // 设置默认日志等级为INFO
        this.logLevel = (_a = options.logLevel) !== null && _a !== void 0 ? _a : const_1.SSOLogLevel.INFO;
        try {
            if (env_1.ENV.isBrowser) {
                var debugValue = sessionStorage && sessionStorage.getItem(const_1.SSO_WEB_OIDC_DEBUG_KEY);
                // 如果sessionStorage中有调试标志，则设置为DEBUG等级
                if (debugValue) {
                    this.logLevel = const_1.SSOLogLevel.DEBUG;
                }
            }
        }
        catch (error) {
            // 静默处理初始化错误
        }
    }
    SSOLogger.prototype.info = function (tag) {
        var _a;
        var args = [];
        for (var _i = 1; _i < arguments.length; _i++) {
            args[_i - 1] = arguments[_i];
        }
        if (this.logLevel >= const_1.SSOLogLevel.INFO) {
            var processedArgs = processLogArgs(args, const_1.SSOLogLevel.INFO);
            (_a = this.output).log.apply(_a, __spreadArray([this.getLogPrefix('INFO', tag)], processedArgs, false));
        }
    };
    SSOLogger.prototype.warn = function (tag) {
        var _a;
        var args = [];
        for (var _i = 1; _i < arguments.length; _i++) {
            args[_i - 1] = arguments[_i];
        }
        if (this.logLevel >= const_1.SSOLogLevel.WARN) {
            var processedArgs = processLogArgs(args, const_1.SSOLogLevel.WARN);
            (_a = this.output).log.apply(_a, __spreadArray([this.getLogPrefix('WARN', tag)], processedArgs, false));
        }
    };
    SSOLogger.prototype.error = function (tag) {
        var _a;
        var args = [];
        for (var _i = 1; _i < arguments.length; _i++) {
            args[_i - 1] = arguments[_i];
        }
        if (this.logLevel >= const_1.SSOLogLevel.ERROR) {
            var processedArgs = processLogArgs(args, const_1.SSOLogLevel.ERROR);
            (_a = this.output).log.apply(_a, __spreadArray([this.getLogPrefix('ERROR', tag)], processedArgs, false));
        }
    };
    SSOLogger.prototype.debug = function (tag) {
        var _a;
        var args = [];
        for (var _i = 1; _i < arguments.length; _i++) {
            args[_i - 1] = arguments[_i];
        }
        if (this.logLevel >= const_1.SSOLogLevel.DEBUG) {
            var processedArgs = processLogArgs(args, const_1.SSOLogLevel.DEBUG);
            (_a = this.output).log.apply(_a, __spreadArray([this.getLogPrefix('DEBUG', tag)], processedArgs, false));
        }
    };
    // 销毁Logger，清理资源
    SSOLogger.prototype.destroy = function () {
        if (this.output && this.output.destroy) {
            this.output.destroy();
        }
    };
    SSOLogger.prototype.updateLogLevel = function (logLevel) {
        this.logLevel = logLevel;
    };
    SSOLogger.prototype.getLogPrefix = function (type, TAG) {
        if (this.packageName) {
            return "[".concat(type, "] ").concat(this.packageName, " - ").concat(TAG ? "[".concat(TAG, "]") : '');
        }
        return "[".concat(type, "] ").concat(TAG ? "[".concat(TAG, "]") : '');
    };
    return SSOLogger;
}());
exports.default = SSOLogger;
