// @mtfe/sso-web-oidc-base/lib/utils/utils.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.buildTaskTagName = exports.deduplicateTask = exports.getLevelOnePermissionApplyUrl = exports.createAuthorizationHeader = exports.getLogoutUrl = exports.getLoginUrl = exports.getOriginalUrl = exports.buildResult = exports.getRealOrigin = exports.getFixedSearchParams = void 0;
exports.generateState = generateState;
var const_1 = require("../config/const");
var query_string_1 = __importDefault(require("query-string"));
var crypto_1 = require("./crypto");
var env_1 = require("./env");
var common_1 = require("./common");
var typeUtil_1 = require("./typeUtil");
/**
 * 获取固定查询参数对象（不要随便添加参数，甚至包括登录参数也不要随便添加）
 * @param moreParams
 * @returns 参数对象
 */
var getFixedSearchParams = function (_a) {
    var _b;
    var sdkVersion = _a.sdkVersion, locale = _a.locale, mode = _a.mode;
    var result = (_b = {},
        _b[const_1.LOGIN_PROTOCOL_KEY] = const_1.DEFAULT_LOGIN_PROTOCOL,
        _b);
    if (sdkVersion) {
        result.sdk = sdkVersion;
    }
    if (locale) {
        result.locale = locale;
    }
    if (mode) {
        result.mode = mode;
    }
    return result;
};
exports.getFixedSearchParams = getFixedSearchParams;
/**
 * 获取实际使用域名
 * @param originMap 域名映射
 * @param configs 登录配置
 * @returns 域名
 */
var getRealOrigin = function (originMap, configs) {
    var accessEnv = configs.accessEnv, swimlane = configs.swimlane;
    if (accessEnv === const_1.SSOAccessEnvType.test && swimlane) {
        var raw = originMap[accessEnv];
        var _a = raw.split('://'), first = _a[0], second = _a[1];
        return "".concat(first, "://").concat(swimlane, "-sl-").concat(second);
    }
    return originMap[accessEnv];
};
exports.getRealOrigin = getRealOrigin;
/**
 * 获取返回结构体
 * @param code 错误码
 * @param data 其他数据
 * @param msg 错误消息
 * @returns 返回结构体
 */
var buildResult = function (code, data, msg) {
    return {
        code: code,
        data: data,
        msg: msg,
    };
};
exports.buildResult = buildResult;
/**
 * 生成state参数
 * @returns state值
 */
function generateState() {
    return Math.random().toString(32).slice(2) + Math.random().toString(32).slice(2);
}
/**
 * 获取当前页面的相对路径
 * @returns 当前页面的相对路径（包含查询参数和哈希值）
 */
var getOriginalUrl = function (configs) {
    if (configs.mode === const_1.SSOWorkModel.cli || !env_1.ENV.isBrowser)
        return '/';
    var _a = location.pathname, pathname = _a === void 0 ? '/' : _a, _b = location.search, search = _b === void 0 ? '' : _b, _c = location.hash, hash = _c === void 0 ? '' : _c;
    return pathname + search + hash;
};
exports.getOriginalUrl = getOriginalUrl;
/**
 * 获取登录链接
 * @param configs SSO登录配置
 * @returns 登录链接
 */
var getLoginUrl = function (_a) {
    var _b;
    var configs = _a.configs, store = _a.store, localPort = _a.localPort, sdkVersion = _a.sdkVersion, customOriginalUrl = _a.customOriginalUrl;
    var clientId = configs.clientId, mode = configs.mode, maxAge = configs.maxAge, secondFactorList = configs.secondFactorList, prompt = configs.prompt, ignoreSLOType = configs.ignoreSLOType;
    // 相对路径
    var originalUrl = customOriginalUrl ? customOriginalUrl : (0, exports.getOriginalUrl)(configs);
    var targetOrigin;
    if (mode === const_1.SSOWorkModel.cli) {
        targetOrigin = "http://localhost:".concat(localPort);
    }
    else {
        targetOrigin = env_1.ENV.isBrowser ? location.origin : '';
    }
    // 回调地址
    var redirectUri = "".concat(targetOrigin + const_1.SSOURLEnum.CALLBACK, "?").concat(query_string_1.default.stringify((_b = {},
        _b[const_1.SSO_ORIGINAL_URL_KEY] = originalUrl,
        _b)));
    var state = generateState();
    var codeVerifier = (0, crypto_1.generateCodeVerifier)();
    // 请求参数
    var searchParams = __assign(__assign({}, (0, exports.getFixedSearchParams)(__assign({ sdkVersion: sdkVersion }, configs))), { scope: const_1.OAUTH_SCOPE, response_type: const_1.OAUTH_RESPONSE_TYPE, client_id: clientId, state: state, code_challenge_method: const_1.OAUTH_PKCE_METHOD, code_challenge: (0, crypto_1.getCodeChallenge)(codeVerifier), redirect_uri: redirectUri });
    if (maxAge && Array.isArray(secondFactorList) && secondFactorList.length) {
        searchParams.max_age = maxAge;
        searchParams.acr_values = secondFactorList.map(common_1.convertSecondFactorToAcr).join(' ');
    }
    if (prompt) {
        searchParams.prompt = prompt;
    }
    if (ignoreSLOType) {
        searchParams.ignore_slo_type = ignoreSLOType;
    }
    var keySuffix = "_".concat((state || '').substring(0, 6));
    store.setState(state, keySuffix);
    store.setVerifier(codeVerifier, keySuffix);
    store.setConfigs(configs, keySuffix);
    return "".concat((0, exports.getRealOrigin)(const_1.SSO_ORIGIN_MAP, configs)).concat(const_1.SSO_OIDC_LOGIN_PATH, "?").concat(query_string_1.default.stringify(searchParams));
};
exports.getLoginUrl = getLoginUrl;
/**
 * 获取登出链接
 * @param configs SSO登录配置
 * @returns 登出链接
 */
var getLogoutUrl = function (_a) {
    var configs = _a.configs, sdkVersion = _a.sdkVersion;
    var clientId = configs.clientId;
    // 相对路径，退出后跳到这个地址
    var originalUrl = (0, exports.getOriginalUrl)(configs);
    var searchParams = __assign(__assign({}, (0, exports.getFixedSearchParams)(__assign({ sdkVersion: sdkVersion }, configs))), { redirect_uri: originalUrl, client_id: clientId });
    if (env_1.ENV.isBrowser)
        return "".concat(location.origin + const_1.SSOURLEnum.LOGOUT, "?").concat(query_string_1.default.stringify(searchParams));
    return '';
};
exports.getLogoutUrl = getLogoutUrl;
/**
 * 创建Authorization Header值
 * @param token token值
 * @returns
 */
var createAuthorizationHeader = function (token) {
    return "Bearer ".concat(token);
};
exports.createAuthorizationHeader = createAuthorizationHeader;
/**
 * 获取一级权限的
 * @param env 环境变量
 * @param targetClientId 目标应用的clientId
 * @returns 申请链接地址
 */
var getLevelOnePermissionApplyUrl = function (accessEnv, targetClientId) {
    var _a;
    var URL_MAP = (_a = {},
        _a[const_1.SSOAccessEnvType.test] = 'https://iam.it.test.sankuai.com/permission-apply',
        _a[const_1.SSOAccessEnvType.product] = 'https://iam.sankuai.com/permission-apply',
        _a);
    var urlParams = {
        activeKey: 'user',
        clientId: targetClientId,
    };
    var url = "".concat(URL_MAP[accessEnv], "?").concat(query_string_1.default.stringify(urlParams));
    return url;
};
exports.getLevelOnePermissionApplyUrl = getLevelOnePermissionApplyUrl;
/**
 * 将一个promise对象转成可查询状态的promise对象
 * @param promise 原始promise对象
 * @returns 可查询的promise对象
 */
function makeQueryablePromise(promise) {
    var isPending = true;
    var isFulfilled = false;
    var isRejected = false;
    var result = promise
        .then(function (v) {
        isPending = false;
        isFulfilled = true;
        return v;
    })
        .catch(function (e) {
        isPending = false;
        isRejected = true;
        throw e;
    });
    result.isPending = function () { return isPending; };
    result.isFulfilled = function () { return isFulfilled; };
    result.isRejected = function () { return isRejected; };
    return result;
}
/**
 * 任务缓存池
 */
var taskCache = {};
/**
 * 如果前一个相同任务处理结束（成功、失败），新任务会直接运行，并加入到任务缓存池中，
 * 否则会使用前一个任务的结果进行返回
 * @param taskTag 任务标记，用来区分任务是否相等
 * @param task 任务运行函数
 * @param lifeTime 新鲜度时间要求（单位：ms）
 * @returns 任务结果
 */
var deduplicateTask = function (taskTag, task, lifeTime) {
    if (lifeTime === void 0) { lifeTime = 100; }
    if (!taskCache[taskTag] ||
        taskCache[taskTag].createdTime + lifeTime < Date.now() ||
        !(taskCache[taskTag].taskResult instanceof Promise) ||
        ((0, typeUtil_1.tsSafeCheckKey)(taskCache[taskTag].taskResult, 'isPending') &&
            !taskCache[taskTag].taskResult.isPending())) {
        var rawTaskResult = task();
        var taskResult = rawTaskResult instanceof Promise ? makeQueryablePromise(rawTaskResult) : rawTaskResult;
        taskCache[taskTag] = {
            createdTime: Date.now(),
            taskResult: taskResult,
        };
        return taskResult;
    }
    console.warn('hint deduplicateTask cache!');
    // 找到taskTag对应的任务缓存，
    // 且在有效期，
    // 且是pending中的promise或非promise
    return taskCache[taskTag].taskResult;
};
exports.deduplicateTask = deduplicateTask;
/**
 * 构建任务标记名
 * @param name 主名称
 * @param arg 参数
 */
var buildTaskTagName = function (name) {
    var args = [];
    for (var _i = 1; _i < arguments.length; _i++) {
        args[_i - 1] = arguments[_i];
    }
    var taskTag = "task__".concat(name, "__");
    if (Array.isArray(args)) {
        args.forEach(function (arg) {
            if (Array.isArray(arg)) {
                taskTag += "@@@".concat(arg.join('#'));
            }
            else if (typeof arg === 'object' && arg !== null && !Array.isArray(arg)) {
                var str = Object.entries(arg)
                    .sort(function (a, b) {
                    if (a[0] === b[0])
                        return 0;
                    if (a[0] < b[0])
                        return -1;
                    return 1;
                })
                    .reduce(function (acc, cur) {
                    return "".concat(acc, "$key$:").concat(cur[0], "$value$:").concat(cur[1]);
                }, '');
                taskTag += "@@@".concat(str);
            }
            else {
                taskTag += "@@@".concat(arg);
            }
        });
    }
    return taskTag;
};
exports.buildTaskTagName = buildTaskTagName;
