// @mtfe/sso-web-oidc-base/lib/core/base.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g = Object.create((typeof Iterator === "function" ? Iterator : Object).prototype);
    return g.next = verb(0), g["throw"] = verb(1), g["return"] = verb(2), typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SSOBaseClient = void 0;
var logger_1 = __importDefault(require("../utils/logger"));
var httpClient_1 = __importDefault(require("../utils/httpClient"));
var utils_1 = require("../utils/utils");
var storage_1 = __importDefault(require("../store/storage"));
var const_1 = require("../config/const");
var env_1 = require("../utils/env");
var common_1 = require("../utils/common");
/**
 * SSO基础客户端
 */
var SSOBaseClient = /** @class */ (function () {
    function SSOBaseClient(_a) {
        var configs = _a.configs, request = _a.request, logger = _a.logger, sdkVersion = _a.sdkVersion;
        var _b, _c, _d;
        if (!const_1.LEGAL_ACCESS_ENV_LIST.includes(configs.accessEnv)) {
            throw new Error("accessEnv must be one of ".concat(const_1.LEGAL_ACCESS_ENV_LIST.join(',')));
        }
        if ((((_b = configs.secondFactorList) === null || _b === void 0 ? void 0 : _b.length) && (0, common_1.isEmpty)(configs.maxAge)) ||
            (!((_c = configs.secondFactorList) === null || _c === void 0 ? void 0 : _c.length) && !(0, common_1.isEmpty)(configs.maxAge))) {
            // eslint-disable-next-line no-console
            console.error("[".concat(sdkVersion, "]: secondFactorList and maxAge must be set together"));
        }
        var secondFactorList;
        if (configs.secondFactorList && configs.secondFactorList.length) {
            configs.secondFactorList.forEach(function (item) {
                if ((0, common_1.isEmpty)(item))
                    return;
                if (!const_1.allSSOSecondFactorTypes[item]) {
                    // eslint-disable-next-line no-console
                    console.error("[".concat(sdkVersion, "]: ").concat(item, " is not valid secondFactor"));
                }
                // 不过滤传入配置项，即使后面新增二次认证类型，SDK也能向前兼容
                if (!secondFactorList)
                    secondFactorList = [];
                secondFactorList.push(item);
            });
        }
        this.sdkVersion = sdkVersion;
        // copy一份配置，避免修改原始配置
        var copyConfig = {
            clientId: configs.clientId,
            accessEnv: configs.accessEnv || const_1.SSOAccessEnvType.test,
            isDebug: configs.isDebug || false,
            mode: configs.mode || const_1.SSOWorkModel.browser, // 默认browser模式
        };
        if (configs.sameSite) {
            copyConfig.sameSite = configs.sameSite;
        }
        if (configs.locale) {
            copyConfig.locale = configs.locale;
        }
        if (configs.mode === const_1.SSOWorkModel.cli) {
            copyConfig.localPortList = ((_d = configs.localPortList) === null || _d === void 0 ? void 0 : _d.length)
                ? configs.localPortList
                : [const_1.CLI_DEFAULT_LOCAL_PORT];
            // 只有cli模式才支持tokenStorage
            copyConfig.tokenStorage = configs.tokenStorage;
        }
        // 兼容泳道测试环境
        if (copyConfig.accessEnv === const_1.SSOAccessEnvType.test && configs.swimlane) {
            copyConfig.swimlane = configs.swimlane;
        }
        if (!(0, common_1.isEmpty)(configs.maxAge) && secondFactorList && secondFactorList.length) {
            copyConfig.maxAge = configs.maxAge;
            copyConfig.secondFactorList = secondFactorList;
        }
        if (configs.prompt) {
            copyConfig.prompt = configs.prompt;
        }
        if (configs.loginTimeout) {
            copyConfig.loginTimeout = configs.loginTimeout;
        }
        if (configs.ignoreSLOType) {
            copyConfig.ignoreSLOType = configs.ignoreSLOType;
        }
        this.initialConfigs = copyConfig;
        this.logger = logger ? logger : new logger_1.default({ mode: this.initialConfigs.mode });
        this.httpClient = new httpClient_1.default({
            configs: this.initialConfigs,
            request: request,
            logger: this.logger,
            sdkVersion: this.sdkVersion,
        });
        this.store = new storage_1.default(this.logger);
        this.logger.debug('initialConfigs', configs);
        this.logger.debug('sdkVersion', this.sdkVersion);
        try {
            this.httpClient.reportConfig();
        }
        catch (e) {
            //
        }
    }
    /**
     * 获取登录链接
     * @param advancedConfig 高级配置
     * @returns 登录链接
     */
    SSOBaseClient.prototype.getLoginUrl = function (advancedConfig) {
        return (0, utils_1.getLoginUrl)({
            configs: __assign({}, this.initialConfigs),
            store: this.store,
            localPort: advancedConfig === null || advancedConfig === void 0 ? void 0 : advancedConfig.localPort,
            sdkVersion: this.sdkVersion,
            customOriginalUrl: advancedConfig === null || advancedConfig === void 0 ? void 0 : advancedConfig.customOriginalUrl,
        });
    };
    /**
     * 获取token等信息
     * 如果AccessToken过期失效，会使用refreshToken刷新accessToken和refreshToken
     * @returns token等信息
     */
    SSOBaseClient.prototype.getToken = function () {
        return __awaiter(this, void 0, void 0, function () {
            var tokenInfo;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.getTokenInfo()];
                    case 1:
                        tokenInfo = _a.sent();
                        return [4 /*yield*/, this.httpClient.getToken(tokenInfo)];
                    case 2: return [2 /*return*/, _a.sent()];
                }
            });
        });
    };
    /**
     * 退出登录
     * 1. 浏览器模式：成功后跳转回当前路径
     * 2. 终端模式，返回结果
     */
    SSOBaseClient.prototype.logout = function () {
        return __awaiter(this, void 0, void 0, function () {
            var mode, access_token;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        mode = this.initialConfigs.mode;
                        if (!(mode === const_1.SSOWorkModel.cli)) return [3 /*break*/, 3];
                        return [4 /*yield*/, this.getTokenInfo()];
                    case 1:
                        access_token = (_a.sent()).access_token;
                        return [4 /*yield*/, this.httpClient.logout(access_token)];
                    case 2: return [2 /*return*/, _a.sent()];
                    case 3:
                        if (env_1.ENV.isBrowser) {
                            location.href = (0, utils_1.getLogoutUrl)({
                                configs: __assign({}, this.initialConfigs),
                                sdkVersion: this.sdkVersion,
                            });
                        }
                        else {
                            this.logger.warn('logout', 'fail: not support in current env');
                        }
                        return [2 /*return*/];
                }
            });
        });
    };
    /**
     * 获取用户信息，失败返回错误信息
     * @returns 用户信息
     */
    SSOBaseClient.prototype.whoami = function () {
        return __awaiter(this, void 0, void 0, function () {
            var access_token;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.getTokenInfo()];
                    case 1:
                        access_token = (_a.sent()).access_token;
                        return [4 /*yield*/, this.httpClient.getUserInfo(access_token)];
                    case 2: return [2 /*return*/, (_a.sent())];
                }
            });
        });
    };
    /**
     * 通过callback回调的code等参数+PKCE参数，换取Token
     * 仅限cli模式
     * @param options callback回调后的参数
     */
    SSOBaseClient.prototype.exchangeTokenByCode = function (options) {
        return __awaiter(this, void 0, void 0, function () {
            var code, state, _a, clientId, sameSite, codeVerifier;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        code = options.code, state = options.state;
                        _a = this.initialConfigs, clientId = _a.clientId, sameSite = _a.sameSite;
                        codeVerifier = this.store.getVerifier();
                        if (!code)
                            return [2 /*return*/, (0, utils_1.buildResult)(const_1.SSOErrorCode.MISS_CODE, null, const_1.CallbackErrorMsg.MISS_CODE)];
                        if (!state || state !== this.store.getState())
                            return [2 /*return*/, (0, utils_1.buildResult)(const_1.SSOErrorCode.STATE_NO_MATCH, null, const_1.CallbackErrorMsg.STATE_NO_MATCH)];
                        if (!codeVerifier)
                            return [2 /*return*/, (0, utils_1.buildResult)(const_1.SSOErrorCode.MISS_CODE_VERIFIER, null, const_1.CallbackErrorMsg.MISS_CODE_VERIFIER)];
                        return [4 /*yield*/, this.httpClient.exchangeTokenByCode({
                                code: code,
                                state: state,
                                clientId: clientId,
                                sameSite: sameSite,
                                codeVerifier: codeVerifier,
                            })];
                    case 1: 
                    // 发起HTTP请求
                    return [2 /*return*/, _b.sent()];
                }
            });
        });
    };
    /**
     * 获取处理过的SSO登录配置
     * @returns 处理过的登录配置
     */
    SSOBaseClient.prototype.getConfigs = function () {
        return __assign({}, this.initialConfigs);
    };
    /**
     * 获取Token存储器的Token信息
     * @returns Token对象
     */
    SSOBaseClient.prototype.getTokenInfo = function () {
        return __awaiter(this, void 0, void 0, function () {
            var e_1;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        _a.trys.push([0, 3, , 4]);
                        if (!(this.initialConfigs.tokenStorage && this.initialConfigs.tokenStorage.get)) return [3 /*break*/, 2];
                        return [4 /*yield*/, this.initialConfigs.tokenStorage.get()];
                    case 1: return [2 /*return*/, (_a.sent()) || {}];
                    case 2: return [3 /*break*/, 4];
                    case 3:
                        e_1 = _a.sent();
                        this.logger.error('getTokenInfo', 'fail', e_1);
                        return [3 /*break*/, 4];
                    case 4: return [2 /*return*/, {}];
                }
            });
        });
    };
    /**
     * 保存Token信息到Token存储器
     * @param tokenInfo Token信息
     */
    SSOBaseClient.prototype.saveTokenInfo = function (tokenInfo) {
        return __awaiter(this, void 0, void 0, function () {
            var e_2;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        _a.trys.push([0, 3, , 4]);
                        if (!(tokenInfo && this.initialConfigs.tokenStorage && this.initialConfigs.tokenStorage.set)) return [3 /*break*/, 2];
                        return [4 /*yield*/, this.initialConfigs.tokenStorage.set(__assign(__assign({}, tokenInfo), { modified_at: Number(new Date()) }))];
                    case 1:
                        _a.sent();
                        this.logger.debug('saveTokenInfo', 'success');
                        _a.label = 2;
                    case 2: return [3 /*break*/, 4];
                    case 3:
                        e_2 = _a.sent();
                        this.logger.error('saveTokenInfo', 'fail', e_2);
                        return [3 /*break*/, 4];
                    case 4: return [2 /*return*/];
                }
            });
        });
    };
    /**
     * 换取Token
     * @param targetClientIdList 目标ClientId数组
     * @returns TokenExchange结果
     */
    SSOBaseClient.prototype.tokenExchange = function (targetClientIdList) {
        return __awaiter(this, void 0, void 0, function () {
            var _a, access_token, token_exchange_attestation, _b, clientId, mode, params, result, getTokenResult;
            return __generator(this, function (_c) {
                switch (_c.label) {
                    case 0: return [4 /*yield*/, this.getTokenInfo()];
                    case 1:
                        _a = _c.sent(), access_token = _a.access_token, token_exchange_attestation = _a.token_exchange_attestation;
                        _b = this.initialConfigs, clientId = _b.clientId, mode = _b.mode;
                        params = {
                            clientId: clientId,
                            audience: targetClientIdList,
                        };
                        if (mode === const_1.SSOWorkModel.cli) {
                            if (!targetClientIdList ||
                                !targetClientIdList.length ||
                                !token_exchange_attestation ||
                                !access_token) {
                                return [2 /*return*/, (0, utils_1.buildResult)(const_1.SSOErrorCode.MISS_PARAMS, null, 'TokenExchange fail: miss params')];
                            }
                            params.token_exchange_attestation = token_exchange_attestation;
                            params.access_token = access_token;
                        }
                        if (mode === const_1.SSOWorkModel.browser && (!targetClientIdList || !targetClientIdList.length)) {
                            return [2 /*return*/, (0, utils_1.buildResult)(const_1.SSOErrorCode.MISS_PARAMS, null, 'TokenExchange fail: miss targetClientIdList')];
                        }
                        return [4 /*yield*/, this.httpClient.tokenExchange(params)];
                    case 2:
                        result = _c.sent();
                        if (!(result.code === const_1.SSOErrorCode.TOKEN_EXCHANGE_FAIL)) return [3 /*break*/, 5];
                        this.logger.debug('tokenExchange', 'tokenExchange fail, retrying');
                        return [4 /*yield*/, this.getToken()];
                    case 3:
                        getTokenResult = _c.sent();
                        if (!(getTokenResult.code === const_1.SSOErrorCode.SUCCESS)) return [3 /*break*/, 5];
                        return [4 /*yield*/, this.httpClient.tokenExchange(params)];
                    case 4:
                        // 重新调用一遍
                        result = _c.sent();
                        _c.label = 5;
                    case 5: return [2 /*return*/, result];
                }
            });
        });
    };
    /**
     * 上报用户点击类型
     * @param args 上报类型参数
     * @returns
     */
    SSOBaseClient.prototype.reportType = function () {
        var _a;
        var args = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            args[_i] = arguments[_i];
        }
        return (_a = this.httpClient).reportType.apply(_a, args);
    };
    return SSOBaseClient;
}());
exports.SSOBaseClient = SSOBaseClient;
