// @mtfe/sso-web-oidc-base/lib/store/storage.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
var env_1 = require("../utils/env");
var const_1 = require("./const");
var SSOStorage = /** @class */ (function () {
    function SSOStorage(logger) {
        this.cache = {};
        this.logger = logger;
    }
    /**
     * дҝқеӯҳstateеҖј
     * @param state stateеҖј
     */
    SSOStorage.prototype.setState = function (state, keySuffix) {
        this.setStorage(const_1.SSO_STATE_KEY, state, { keySuffix: keySuffix });
    };
    /**
     * !!!д»…йҷҗcliжЁЎејҸ
     * иҜ»еҸ–stateеҖј
     * @returns stateеҖј
     */
    SSOStorage.prototype.getState = function () {
        return this.getStorage(const_1.SSO_STATE_KEY);
    };
    /**
     * дҝқеӯҳPKCE code_verifierеҖј
     * @param verifier PKCE verifierеҖј
     */
    SSOStorage.prototype.setVerifier = function (verifier, keySuffix) {
        this.setStorage(const_1.SSO_CODE_VERIFIER_KEY, verifier, { keySuffix: keySuffix });
    };
    /**
     * !!!д»…йҷҗcliжЁЎејҸ
     * иҜ»еҸ–PKCE code_verifierеҖј
     * @returns PKCE verifierеҖј
     */
    SSOStorage.prototype.getVerifier = function () {
        return this.getStorage(const_1.SSO_CODE_VERIFIER_KEY);
    };
    /**
     * дҝқеӯҳSSO web config
     * @param configs SSO web config
     */
    SSOStorage.prototype.setConfigs = function (configs, keySuffix) {
        var value = '{}';
        var realConfigs = __assign({}, (configs || {}));
        if (configs.sameSite) {
            realConfigs.samesite = configs.sameSite;
        }
        try {
            value = JSON.stringify(__assign(__assign({}, realConfigs), { createdTime: Number(new Date()) }));
        }
        catch (error) {
            this.logger.error('setConfigs', 'fail', realConfigs, error);
        }
        this.setStorage(const_1.SSO_WEB_CONFIG_KEY, value, { keySuffix: keySuffix });
    };
    // /**
    //  * иҠұеҺ»SSO web config
    //  * @returns SSO web config
    //  */
    // public getConfigs() {
    //   try {
    //     const rawValue = this.getStorage(SSO_WEB_CONFIG_KEY);
    //     if (!rawValue) return undefined;
    //     return JSON.parse(rawValue);
    //   } catch (error) {
    //     this.logger.error('getConfigs', 'fail', error);
    //     return {};
    //   }
    // }
    /**
     * дҝқеӯҳkeyгҖҒvalue
     * @param key keyеҖј
     * @param value valueеҖј
     */
    SSOStorage.prototype.setStorage = function (key, value, options) {
        if (!key)
            return;
        // е…ҲдҝқеӯҳеҲ°session storage
        try {
            if (env_1.ENV.isBrowser && sessionStorage && sessionStorage.setItem) {
                var realKey = key;
                if (options && options.keySuffix) {
                    realKey += "".concat(options.keySuffix);
                }
                sessionStorage.setItem(realKey, value);
                this.logger.debug("Try to save key ".concat(realKey, " and value ").concat(value, " in sessionStorage"));
                if (sessionStorage.getItem(realKey) !== value) {
                    this.logger.error('Saving sessionStorage failed');
                }
            }
        }
        catch (error) {
            this.logger.error('setSessionStorage', 'fail', key, value, error);
        }
        // еҶҚдҝқеӯҳеҲ°local storageпјҲе…ңеә•пјү
        try {
            if (env_1.ENV.isBrowser && localStorage && localStorage.setItem) {
                localStorage.setItem(key, value);
            }
        }
        catch (error) {
            this.logger.error('setLocalStorage', 'fail', key, value, error);
        }
        // дҝқеӯҳеҲ°cacheдёӯ
        this.cache[key] = value;
    };
    /**
     * иҜ»еҸ–keyеҜ№еә”зҡ„valueеҖј
     * @param key keyеҖј
     * @returns valueеҖј
     */
    SSOStorage.prototype.getStorage = function (key) {
        if (!key)
            return null;
        // е…Ҳд»Һsession storageиҜ»еҸ–
        try {
            if (env_1.ENV.isBrowser && sessionStorage && sessionStorage.getItem) {
                var value = sessionStorage.getItem(key);
                if (value)
                    return value;
            }
        }
        catch (error) {
            this.logger.error('getSsessionStorage', 'fail', key, error);
        }
        // еҶҚд»Һlocal storageпјҲе…ңеә•пјүиҜ»еҸ–
        try {
            if (env_1.ENV.isBrowser && localStorage && localStorage.getItem) {
                var value = localStorage.getItem(key);
                if (value)
                    return value;
            }
        }
        catch (error) {
            this.logger.error('getLocalStorage', 'fail', key, error);
        }
        // жңҖз»Ҳд»ҺcacheдёӯиҜ»еҸ–
        return this.cache[key];
    };
    return SSOStorage;
}());
exports.default = SSOStorage;
