// @mtfe/sso-web-oidc-cli/lib/index.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g = Object.create((typeof Iterator === "function" ? Iterator : Object).prototype);
    return g.next = verb(0), g["throw"] = verb(1), g["return"] = verb(2), typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SSOCliClient = exports.SSOIgnoreSLOType = exports.SSOPromptEnum = exports.SSOSecondFactorEnum = exports.openURL = exports.SSOLogLevel = exports.SSOErrorCode = exports.SSOLogger = exports.SSOLanguage = exports.SSOSameSiteType = exports.SSOWorkModel = exports.SSOAccessEnvType = void 0;
var node_fetch_1 = __importDefault(require("node-fetch"));
var proper_lockfile_1 = __importDefault(require("proper-lockfile"));
var sso_web_oidc_base_1 = require("@mtfe/sso-web-oidc-base");
Object.defineProperty(exports, "SSOAccessEnvType", { enumerable: true, get: function () { return sso_web_oidc_base_1.SSOAccessEnvType; } });
Object.defineProperty(exports, "SSOWorkModel", { enumerable: true, get: function () { return sso_web_oidc_base_1.SSOWorkModel; } });
Object.defineProperty(exports, "SSOSameSiteType", { enumerable: true, get: function () { return sso_web_oidc_base_1.SSOSameSiteType; } });
Object.defineProperty(exports, "SSOLanguage", { enumerable: true, get: function () { return sso_web_oidc_base_1.SSOLanguage; } });
Object.defineProperty(exports, "SSOLogger", { enumerable: true, get: function () { return sso_web_oidc_base_1.SSOLogger; } });
Object.defineProperty(exports, "SSOErrorCode", { enumerable: true, get: function () { return sso_web_oidc_base_1.SSOErrorCode; } });
Object.defineProperty(exports, "SSOLogLevel", { enumerable: true, get: function () { return sso_web_oidc_base_1.SSOLogLevel; } });
Object.defineProperty(exports, "SSOSecondFactorEnum", { enumerable: true, get: function () { return sso_web_oidc_base_1.SSOSecondFactorEnum; } });
Object.defineProperty(exports, "SSOPromptEnum", { enumerable: true, get: function () { return sso_web_oidc_base_1.SSOPromptEnum; } });
Object.defineProperty(exports, "SSOIgnoreSLOType", { enumerable: true, get: function () { return sso_web_oidc_base_1.SSOIgnoreSLOType; } });
var server_1 = require("./utils/server");
var version_1 = require("./version");
var crypto_1 = require("./utils/crypto");
var fs_1 = require("./utils/fs");
var utils_1 = require("./utils/utils");
var open_1 = __importDefault(require("open"));
exports.openURL = open_1.default;
var tokenStorage_1 = require("./utils/tokenStorage");
var SDK_NAME = 'SSOWebOIDCCli';
var SSOCliClient = /** @class */ (function () {
    function SSOCliClient(configs) {
        // 必须设置第二个参数
        this.logger = new sso_web_oidc_base_1.SSOLogger({
            mode: sso_web_oidc_base_1.SSOWorkModel.cli,
            logLevel: (configs === null || configs === void 0 ? void 0 : configs.isDebug) ? sso_web_oidc_base_1.SSOLogLevel.DEBUG : undefined,
            output: configs === null || configs === void 0 ? void 0 : configs.logOutputer,
        });
        var tag = configs === null || configs === void 0 ? void 0 : configs.tag;
        var realTokenStorage = (configs === null || configs === void 0 ? void 0 : configs.tokenStorage)
            ? configs.tokenStorage
            : new tokenStorage_1.DefaultMemoryTokenStorage();
        this.baseClient = new sso_web_oidc_base_1.SSOBaseClient({
            configs: __assign(__assign({}, configs), { mode: sso_web_oidc_base_1.SSOWorkModel.cli, tokenStorage: realTokenStorage }),
            // @ts-ignore
            request: function () {
                var args = [];
                for (var _i = 0; _i < arguments.length; _i++) {
                    args[_i] = arguments[_i];
                }
                return node_fetch_1.default.apply(void 0, args);
            },
            logger: this.logger,
            sdkVersion: "".concat(SDK_NAME).concat(tag ? "-".concat(tag) : '', "_").concat((0, version_1.getPkgVersion)()),
        });
    }
    /**
     * 登录方法
     * 调用后，如果已有Token信息，会自动验证并返回有效的Token信息，否则会触发登录流程
     * 1. 成功后返回accessToken、refreshToken等信息
     * 2. 当出现异常情况时，如网络错误、端口占用、登录超时等，抛出异常
     *
     * @param options 登录选项
     * @param options.abortSignal 取消信号，用于取消登录流程
     * @returns 成功后返回accessToken、refreshToken等信息，其他情况返回诊断信息(仅供诊断调试用)
     */
    SSOCliClient.prototype.login = function (options) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.refreshOrLogin({ abortSignal: options === null || options === void 0 ? void 0 : options.abortSignal })];
                    case 1: return [2 /*return*/, _a.sent()];
                }
            });
        });
    };
    /**
     * 续期方法、获取token方法，不会触发登录流程
     * 1. 当accessToken过期时，调用该方法可以进行续期
     * 2. 手动获取token等信息，即使登录过期，也不会触发登录流程
     * @param options 刷新选项
     * @param options.abortSignal 取消信号，用于取消刷新流程
     * @returns 成功后返回最新的access、refreshToken等信息，其他情况抛出异常
     */
    SSOCliClient.prototype.refresh = function (options) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.refreshOrLogin({ onlyRefresh: true, abortSignal: options === null || options === void 0 ? void 0 : options.abortSignal })];
                    case 1: return [2 /*return*/, _a.sent()];
                }
            });
        });
    };
    /**
     * 退出登录
     * @return 返回一个Promise
     */
    SSOCliClient.prototype.logout = function () {
        return this.baseClient.logout();
    };
    /**
     * 获取用户信息，失败返回错误信息
     * @returns 用户信息
     */
    SSOCliClient.prototype.whoami = function () {
        return this.baseClient.whoami();
    };
    /**
     * 换取其他应用的AccessToken
     * @param targetClientId 其他应用的ClientId
     * @returns TokenExchange结果
     */
    SSOCliClient.prototype.tokenExchange = function (targetClientId) {
        return __awaiter(this, void 0, void 0, function () {
            var loginConfig, clientId, TAG;
            var _this = this;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        loginConfig = this.baseClient.getConfigs();
                        clientId = loginConfig.clientId;
                        TAG = 'tokenExchange';
                        return [4 /*yield*/, (0, utils_1.retryWithBackoff)(function () { return __awaiter(_this, void 0, void 0, function () {
                                var lockDir, targetClientIdLock, lockRelease, reuslt;
                                return __generator(this, function (_a) {
                                    switch (_a.label) {
                                        case 0: return [4 /*yield*/, (0, fs_1.ensureLockDir)(this.baseClient.getConfigs())];
                                        case 1:
                                            lockDir = _a.sent();
                                            targetClientIdLock = "TE_".concat((0, crypto_1.getHash)(clientId), "_").concat((0, crypto_1.getHash)(targetClientId));
                                            return [4 /*yield*/, proper_lockfile_1.default.lockSync(targetClientIdLock, {
                                                    realpath: false,
                                                    lockfilePath: "".concat(lockDir, "/").concat(targetClientIdLock),
                                                })];
                                        case 2:
                                            lockRelease = _a.sent();
                                            _a.label = 3;
                                        case 3:
                                            _a.trys.push([3, , 5, 6]);
                                            return [4 /*yield*/, this.baseClient.tokenExchange(targetClientId ? [targetClientId] : [])];
                                        case 4:
                                            reuslt = _a.sent();
                                            this.logger.debug(TAG, 'success');
                                            return [2 /*return*/, reuslt];
                                        case 5:
                                            lockRelease();
                                            return [7 /*endfinally*/];
                                        case 6: return [2 /*return*/];
                                    }
                                });
                            }); }, this.logger, TAG)];
                    case 1: return [2 /*return*/, _a.sent()];
                }
            });
        });
    };
    SSOCliClient.prototype.checkAndSave = function (newTokenInfo) {
        return __awaiter(this, void 0, void 0, function () {
            var oldTokenInfo, tokenInfo;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (!newTokenInfo)
                            return [2 /*return*/];
                        return [4 /*yield*/, this.baseClient.getTokenInfo()];
                    case 1:
                        oldTokenInfo = (_a.sent()) || {};
                        if (!(oldTokenInfo.access_token !== newTokenInfo.access_token ||
                            oldTokenInfo.refresh_token !== newTokenInfo.refresh_token)) return [3 /*break*/, 3];
                        tokenInfo = __assign(__assign({}, newTokenInfo), { token_exchange_attestation: newTokenInfo.token_exchange_attestation || oldTokenInfo.token_exchange_attestation });
                        return [4 /*yield*/, this.baseClient.saveTokenInfo(tokenInfo)];
                    case 2:
                        _a.sent();
                        _a.label = 3;
                    case 3: return [2 /*return*/];
                }
            });
        });
    };
    SSOCliClient.prototype.refreshOrLogin = function (options) {
        return __awaiter(this, void 0, void 0, function () {
            var loginConfig, clientId, loginTimeout, TAG, abortSignal;
            var _this = this;
            var _a;
            return __generator(this, function (_b) {
                // 检查传入的 abortSignal 是否已经被取消
                if ((_a = options === null || options === void 0 ? void 0 : options.abortSignal) === null || _a === void 0 ? void 0 : _a.aborted) {
                    throw (0, sso_web_oidc_base_1.buildResult)(sso_web_oidc_base_1.SSOErrorCode.MANUAL_CANCEL, null, 'refreshOrLogin操作已被取消');
                }
                loginConfig = this.baseClient.getConfigs();
                clientId = loginConfig.clientId, loginTimeout = loginConfig.loginTimeout;
                TAG = 'refreshOrLogin';
                abortSignal = options === null || options === void 0 ? void 0 : options.abortSignal;
                // 创建并存储当前的 Promise
                return [2 /*return*/, (0, utils_1.retryWithBackoff)(function () { return __awaiter(_this, void 0, void 0, function () {
                        var clientIdLock, lockDir, lockRelease, _a, code, msg, data, oauthResult;
                        return __generator(this, function (_b) {
                            switch (_b.label) {
                                case 0:
                                    // 在每个重要步骤前检查是否被取消
                                    if (abortSignal === null || abortSignal === void 0 ? void 0 : abortSignal.aborted) {
                                        throw (0, sso_web_oidc_base_1.buildResult)(sso_web_oidc_base_1.SSOErrorCode.MANUAL_CANCEL, null, '登录或刷新操作已被取消');
                                    }
                                    clientIdLock = "login_".concat((0, crypto_1.getHash)(clientId));
                                    return [4 /*yield*/, (0, fs_1.ensureLockDir)(loginConfig)];
                                case 1:
                                    lockDir = _b.sent();
                                    return [4 /*yield*/, proper_lockfile_1.default.lockSync(clientIdLock, {
                                            realpath: false,
                                            lockfilePath: "".concat(lockDir, "/").concat(clientIdLock),
                                        })];
                                case 2:
                                    lockRelease = _b.sent();
                                    _b.label = 3;
                                case 3:
                                    _b.trys.push([3, , 14, 15]);
                                    // 再次检查是否被取消
                                    if (abortSignal === null || abortSignal === void 0 ? void 0 : abortSignal.aborted) {
                                        throw (0, sso_web_oidc_base_1.buildResult)(sso_web_oidc_base_1.SSOErrorCode.MANUAL_CANCEL, null, '登录或刷新操作已被取消');
                                    }
                                    return [4 /*yield*/, this.baseClient.getToken()];
                                case 4:
                                    _a = _b.sent(), code = _a.code, msg = _a.msg, data = _a.data;
                                    if (!(code === sso_web_oidc_base_1.SSOErrorCode.UNAUTHORIZED)) return [3 /*break*/, 10];
                                    this.logger.debug(TAG, 'unauthorized', { code: code, data: data, msg: msg });
                                    if (!(options && options.onlyRefresh)) return [3 /*break*/, 5];
                                    throw (0, sso_web_oidc_base_1.buildResult)(code, data, msg);
                                case 5:
                                    // 检查是否被取消
                                    if (abortSignal === null || abortSignal === void 0 ? void 0 : abortSignal.aborted) {
                                        throw (0, sso_web_oidc_base_1.buildResult)(sso_web_oidc_base_1.SSOErrorCode.MANUAL_CANCEL, null, '登录或刷新操作已被取消');
                                    }
                                    return [4 /*yield*/, (0, server_1.waitForOAuthCallback)({
                                            logger: this.logger,
                                            baseClient: this.baseClient,
                                            timeout: loginTimeout,
                                            abortSignal: abortSignal,
                                        })];
                                case 6:
                                    oauthResult = _b.sent();
                                    if (!(oauthResult && oauthResult.code === sso_web_oidc_base_1.SSOErrorCode.SUCCESS && oauthResult.data)) return [3 /*break*/, 8];
                                    return [4 /*yield*/, this.baseClient.saveTokenInfo(oauthResult.data)];
                                case 7:
                                    _b.sent();
                                    return [2 /*return*/, oauthResult.data];
                                case 8: throw oauthResult;
                                case 9: return [3 /*break*/, 13];
                                case 10:
                                    if (!(code !== sso_web_oidc_base_1.SSOErrorCode.SUCCESS || !data)) return [3 /*break*/, 11];
                                    this.logger.error(TAG, 'fail', { code: code, data: data, msg: msg });
                                    throw (0, sso_web_oidc_base_1.buildResult)(code, data, msg);
                                case 11:
                                    this.logger.debug(TAG, 'success');
                                    // 成功获取到新的Token信息
                                    return [4 /*yield*/, this.checkAndSave(data)];
                                case 12:
                                    // 成功获取到新的Token信息
                                    _b.sent();
                                    return [2 /*return*/, data];
                                case 13: return [3 /*break*/, 15];
                                case 14:
                                    lockRelease();
                                    return [7 /*endfinally*/];
                                case 15: return [2 /*return*/];
                            }
                        });
                    }); }, this.logger, TAG)];
            });
        });
    };
    return SSOCliClient;
}());
exports.SSOCliClient = SSOCliClient;
