// @mtfe/sso-web-oidc-cli/lib/utils/server.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g = Object.create((typeof Iterator === "function" ? Iterator : Object).prototype);
    return g.next = verb(0), g["throw"] = verb(1), g["return"] = verb(2), typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.waitForOAuthCallback = void 0;
var sso_web_oidc_base_1 = require("@mtfe/sso-web-oidc-base");
var http_1 = require("http");
var const_1 = require("./const");
var renderHtml_1 = __importDefault(require("./renderHtml"));
var string_1 = require("./string");
var open_1 = __importDefault(require("open"));
var random_1 = require("./random");
// 全局活跃服务器集合，用于信号处理时统一清理
var activeServers = new Set();
var waitForOAuthCallback = function (_a) {
    var logger = _a.logger, baseClient = _a.baseClient, _b = _a.timeout, timeout = _b === void 0 ? const_1.SSO_LOGIN_TIMEOUT_IN_SECONDS : _b, abortSignal = _a.abortSignal;
    /**
     * 打开登录链接
     * @param currentPort 当前使用端口
     */
    var openLoginUrl = function (currentPort) {
        var loginUrl = baseClient.getLoginUrl({ localPort: currentPort });
        logger.debug('openLoginUrl', loginUrl);
        // eslint-disable-next-line @typescript-eslint/no-floating-promises
        (0, open_1.default)(loginUrl);
    };
    // 检查是否已经被取消
    if (abortSignal === null || abortSignal === void 0 ? void 0 : abortSignal.aborted) {
        return Promise.reject((0, sso_web_oidc_base_1.buildResult)(sso_web_oidc_base_1.SSOErrorCode.MANUAL_CANCEL, null, 'OAuth流程已被取消'));
    }
    var isCallbackSuccess = false;
    var waitingTimer;
    var realTimeout = Number(timeout) || const_1.SSO_LOGIN_TIMEOUT_IN_SECONDS;
    // 将异步逻辑提取到单独的函数中
    var executeOAuthFlow = function () {
        var _a = baseClient.getConfigs(), _b = _a.localPortList, localPortList = _b === void 0 ? [] : _b, clientId = _a.clientId;
        var currentRetryCount = 0;
        if (!localPortList.length) {
            throw (0, sso_web_oidc_base_1.buildResult)(sso_web_oidc_base_1.SSOErrorCode.NO_AVALIABLE_PORT, null, '无可用端口配置');
        }
        var getPort = function (index) {
            var portSize = localPortList.length;
            return localPortList[index % portSize];
        };
        var MAX_TRY_LIMIT = Math.max(const_1.DEFAULT_MAX_TRY_LIMIT, (localPortList === null || localPortList === void 0 ? void 0 : localPortList.length) || 0);
        var currentServer;
        // 优雅关闭服务器的函数
        var gracefulShutdown = function (signal) {
            logger.debug('gracefulShutdown', "Received ".concat(signal, ", gracefully shutting down server..."));
            // 关闭所有活跃的服务器
            activeServers.forEach(function (server) {
                server === null || server === void 0 ? void 0 : server.close(function (err) {
                    if (err) {
                        logger.error('gracefulShutdown', 'Error during server shutdown:', err);
                    }
                    else {
                        logger.debug('gracefulShutdown', 'Server closed successfully');
                    }
                });
            });
            process.exit(0);
        };
        // 监听退出信号（允许重复监听，影响有限）
        process.once('SIGINT', function () { return gracefulShutdown('SIGINT'); }); // Ctrl+C
        process.once('SIGTERM', function () { return gracefulShutdown('SIGTERM'); }); // 终止信号
        // 返回一个 Promise 来处理 OAuth 回调
        return new Promise(function (resolve, reject) {
            var isResolved = false; // 防止重复调用 resolve/reject
            // 清理函数
            var cleanup = function () {
                try {
                    // 移除取消监听器
                    if (abortSignal) {
                        abortSignal.removeEventListener('abort', handleAbort);
                    }
                    // 清理定时器
                    if (waitingTimer) {
                        clearTimeout(waitingTimer);
                    }
                    // 关闭服务器
                    if (currentServer) {
                        currentServer.close(function (err) {
                            if (err) {
                                logger.error('cleanup', 'Error during server shutdown:', err);
                            }
                            else {
                                logger.debug('cleanup', 'Server closed successfully after cancellation');
                            }
                        });
                        activeServers.delete(currentServer);
                    }
                }
                catch (error) {
                    logger.error('cleanup', 'Error during cleanup:', error);
                }
            };
            /**
             * 包装原始的 resolve 和 reject 以确保清理
             * wrappedResolve、wrappedReject是结束终点
             */
            var wrappedResolve = function (value) {
                if (isResolved)
                    return; // 防止重复调用
                isResolved = true;
                cleanup();
                resolve(value);
            };
            var wrappedReject = function (reason) {
                if (isResolved)
                    return; // 防止重复调用
                isResolved = true;
                cleanup();
                reject(reason);
            };
            // 添加取消监听器
            var handleAbort = function () {
                wrappedReject((0, sso_web_oidc_base_1.buildResult)(sso_web_oidc_base_1.SSOErrorCode.MANUAL_CANCEL, null, 'OAuth流程已被取消'));
            };
            // 超时处理函数
            var timeoutHandler = function () {
                wrappedReject((0, sso_web_oidc_base_1.buildResult)(sso_web_oidc_base_1.SSOErrorCode.TIMEOUT, null, "\u767B\u5F55\u6388\u6743\u7B49\u5F85".concat(realTimeout, "\u79D2\uFF0C\u5DF2\u8D85\u65F6...")));
            };
            // 如果提供了 abortSignal，添加监听器
            if (abortSignal) {
                if (abortSignal.aborted) {
                    handleAbort();
                    return;
                }
                abortSignal.addEventListener('abort', handleAbort);
            }
            var tryCreateServer = function (port) {
                return new Promise(function (serverResolve, serverReject) {
                    var currentServerOrigin = "http://localhost:".concat(port);
                    var server = (0, http_1.createServer)(function (req, res) {
                        var parsedUrl = new URL(req.url || '', currentServerOrigin);
                        logger.debug('OAuthCallbackServer', 'received request from ', parsedUrl.pathname);
                        if (parsedUrl.pathname && parsedUrl.pathname === sso_web_oidc_base_1.SSOURLEnum.CALLBACK) {
                            logger.debug('OAuthCallbackServer', "Received OAuth callback: ".concat(req.url));
                            var code = parsedUrl.searchParams.get('code');
                            var receivedClientId = parsedUrl.searchParams.get('client_id');
                            var state = parsedUrl.searchParams.get('state');
                            var originalUrl = parsedUrl.searchParams.get('original-url');
                            logger.debug('OAuthCallbackServer', "Authorization params received:", {
                                code: code,
                                clientId: receivedClientId,
                                state: state,
                                originalUrl: originalUrl,
                            });
                            baseClient
                                .exchangeTokenByCode({ code: code, state: state })
                                .then(function (result) { return __awaiter(void 0, void 0, void 0, function () {
                                var errorCode, errorMsg;
                                var _a, _b;
                                return __generator(this, function (_c) {
                                    // 检查是否已被取消
                                    if (abortSignal === null || abortSignal === void 0 ? void 0 : abortSignal.aborted) {
                                        handleAbort();
                                        return [2 /*return*/];
                                    }
                                    logger.debug('exchangeTokenByCode', 'success', (_b = (_a = result === null || result === void 0 ? void 0 : result.data) === null || _a === void 0 ? void 0 : _a.access_token) === null || _b === void 0 ? void 0 : _b.substring(0, 10));
                                    if (result && result.code === sso_web_oidc_base_1.SSOErrorCode.SUCCESS) {
                                        res.writeHead(200, { 'Content-Type': 'text/html' });
                                        res.end((0, renderHtml_1.default)({
                                            header: 'SSO登录成功',
                                            duration: 3,
                                            isSuccess: true,
                                        }));
                                        wrappedResolve(result); // 这里会 resolve 外层 Promise
                                    }
                                    else {
                                        res.writeHead(400);
                                        errorCode = (result === null || result === void 0 ? void 0 : result.code) || sso_web_oidc_base_1.SSOErrorCode.UNKNOWN_ERROR;
                                        errorMsg = (result === null || result === void 0 ? void 0 : result.msg) || const_1.ErrorMsg.UNKNOWN_ERROR;
                                        res.end((0, string_1.stringConcatenation)(errorCode, ' - ', errorMsg));
                                        wrappedReject((0, sso_web_oidc_base_1.buildResult)(errorCode, null, errorMsg)); // 这里会 reject 外层 Promise
                                    }
                                    return [2 /*return*/];
                                });
                            }); })
                                .catch(function (error) {
                                logger.error('exchangeTokenByCode', 'fail', error);
                                res.writeHead(500);
                                res.end((0, renderHtml_1.default)({
                                    header: '500 Unkown Error',
                                    duration: 3,
                                }));
                                wrappedReject((0, sso_web_oidc_base_1.buildResult)(sso_web_oidc_base_1.SSOErrorCode.UNKNOWN_ERROR, null, (0, string_1.stringConcatenation)(const_1.ErrorMsg.UNKNOWN_ERROR, ' - ', (error && (error.msg || error.message)) || '')));
                            })
                                .finally(function () {
                                isCallbackSuccess = true;
                                if (waitingTimer)
                                    clearTimeout(waitingTimer);
                                // 设置强制关闭的超时时间
                                var forceCloseTimeout = setTimeout(function () {
                                    logger.debug('tryCreateServer', 'Force closing server due to timeout');
                                    activeServers.delete(server);
                                    server.close(function (err) {
                                        if (!err) {
                                            logger.debug('tryCreateServer', 'Server closed successfully');
                                        }
                                    });
                                }, 3000); // 3秒强制关闭超时
                                // 确保响应发送完毕后再关闭服务器
                                res.on('finish', function () {
                                    clearTimeout(forceCloseTimeout); // 清除强制关闭超时
                                    setTimeout(function () {
                                        activeServers.delete(server);
                                        server.close(function (err) {
                                            if (!err) {
                                                logger.debug('tryCreateServer', 'Server closed successfully');
                                            }
                                        });
                                    }, 50); // 减少延迟时间，因为已经等待响应完成
                                });
                            });
                        }
                        else if (parsedUrl.pathname &&
                            parsedUrl.pathname === sso_web_oidc_base_1.SSOURLEnum.CALLBACK_SERVER_WAITING) {
                            res.writeHead(202, { 'Content-Type': 'application/json' });
                            res.end(JSON.stringify((0, sso_web_oidc_base_1.buildResult)(sso_web_oidc_base_1.SSOErrorCode.PROCESSING, { clientId: clientId }, 'Please waiting for oauth complete.')));
                            return;
                        }
                        else {
                            res.writeHead(404);
                            res.end((0, renderHtml_1.default)({ header: '页面不存在', duration: 2 }));
                            return;
                        }
                    });
                    server.on('error', function (err) {
                        var msg = '服务器启动失败';
                        if (err.code === 'EADDRINUSE') {
                            msg = "[http] - ".concat(err.code, " \u7AEF\u53E3 ").concat(port, " \u5DF2\u88AB\u5360\u7528");
                        }
                        else if (err.code === 'EACCES') {
                            msg = "[http] - ".concat(err.code, " \u6CA1\u6709\u6743\u9650\u7ED1\u5B9A\u7AEF\u53E3 ").concat(port);
                        }
                        else if (err.code === 'EADDRNOTAVAIL') {
                            msg = "[http] - ".concat(err.code, " \u5730\u5740\u4E0D\u53EF\u7528");
                        }
                        else {
                            msg = "[http] - ".concat(err.code, " \u670D\u52A1\u5668\u542F\u52A8\u5931\u8D25");
                        }
                        logger.error('OAuthCallbackServer', msg);
                        serverReject((0, sso_web_oidc_base_1.buildResult)(sso_web_oidc_base_1.SSOErrorCode.CREATE_SERVER_ERROR, null, msg));
                    });
                    server.on('listening', function () {
                        var msg = "OAuth callback server started successfully on ".concat(currentServerOrigin);
                        logger.debug('OAuthCallbackServer', msg);
                        openLoginUrl(port);
                        serverResolve((0, sso_web_oidc_base_1.buildResult)(sso_web_oidc_base_1.SSOErrorCode.SUCCESS, { currentPort: port, currentServer: server }, msg));
                    });
                    server.listen(port);
                });
            };
            // 异步重试逻辑
            (function () { return __awaiter(void 0, void 0, void 0, function () {
                var currentPort, createServerResult, error_1;
                return __generator(this, function (_a) {
                    switch (_a.label) {
                        case 0:
                            if (!(currentRetryCount < MAX_TRY_LIMIT)) return [3 /*break*/, 9];
                            // 检查是否被取消
                            if (abortSignal === null || abortSignal === void 0 ? void 0 : abortSignal.aborted) {
                                handleAbort();
                                return [2 /*return*/];
                            }
                            if (!(currentRetryCount >= localPortList.length)) return [3 /*break*/, 2];
                            return [4 /*yield*/, (0, random_1.randomSleep)(100, 200)];
                        case 1:
                            _a.sent();
                            logger.debug('OAuthCallbackServer', '已经尝试所有端口，等待100-200ms后重试');
                            return [3 /*break*/, 4];
                        case 2: return [4 /*yield*/, (0, random_1.randomSleep)(50, 100)];
                        case 3:
                            _a.sent();
                            logger.debug('OAuthCallbackServer', '等待50-100ms后重试');
                            _a.label = 4;
                        case 4:
                            // 再次检查是否被取消（在睡眠后）
                            if (abortSignal === null || abortSignal === void 0 ? void 0 : abortSignal.aborted) {
                                handleAbort();
                                return [2 /*return*/];
                            }
                            _a.label = 5;
                        case 5:
                            _a.trys.push([5, 7, , 8]);
                            currentPort = getPort(currentRetryCount);
                            return [4 /*yield*/, tryCreateServer(currentPort)];
                        case 6:
                            createServerResult = _a.sent();
                            if (createServerResult &&
                                createServerResult.code === sso_web_oidc_base_1.SSOErrorCode.SUCCESS &&
                                createServerResult.data &&
                                createServerResult.data.currentServer) {
                                // eslint-disable-next-line prefer-destructuring
                                currentServer = createServerResult.data.currentServer;
                                // 将服务器添加到活跃服务器集合
                                activeServers.add(currentServer);
                                logger.debug('waitForOAuthCallback', '创建本地服务器成功');
                                // 再次检查是否被取消（创建完服务器后）
                                if (abortSignal === null || abortSignal === void 0 ? void 0 : abortSignal.aborted) {
                                    handleAbort();
                                    return [2 /*return*/];
                                }
                                if (!isCallbackSuccess) {
                                    // 添加授权超时关闭函数
                                    waitingTimer = setTimeout(timeoutHandler, Number(realTimeout) * 1000);
                                }
                                return [2 /*return*/];
                            }
                            logger.debug('OAuthCallbackServer', '创建失败，抛出相关信息', createServerResult);
                            throw createServerResult;
                        case 7:
                            error_1 = _a.sent();
                            logger.debug('OAuthCallbackServer', "\u7B2C".concat(currentRetryCount + 1, "\u6B21\u521B\u5EFA\u670D\u52A1\u5668\u5931\u8D25\uFF0C\u5C1D\u8BD5\u4E0B\u4E00\u4E2A\u7AEF\u53E3").concat(getPort(currentRetryCount + 1)));
                            return [3 /*break*/, 8];
                        case 8:
                            currentRetryCount += 1;
                            return [3 /*break*/, 0];
                        case 9:
                            // 如果循环正常结束（达到最大重试次数），说明无法启动服务器
                            wrappedReject((0, sso_web_oidc_base_1.buildResult)(sso_web_oidc_base_1.SSOErrorCode.RETRY_TOO_MANY_TIME, null, "\u5185\u90E8\u591A\u6B21\u5C1D\u8BD5\u540E\u4ECD\u65E0\u6CD5\u542F\u52A8\u76D1\u542C\u670D\u52A1\u5668"));
                            return [2 /*return*/];
                    }
                });
            }); })().catch(wrappedReject);
        });
    };
    // 直接返回异步函数的结果
    return executeOAuthFlow();
};
exports.waitForOAuthCallback = waitForOAuthCallback;
