// @mtfe/sso-web-oidc-cli/lib/utils/utils.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g = Object.create((typeof Iterator === "function" ? Iterator : Object).prototype);
    return g.next = verb(0), g["throw"] = verb(1), g["return"] = verb(2), typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.builtError = exports.sleep = void 0;
exports.findFirstAvailablePort = findFirstAvailablePort;
exports.retryWithBackoff = retryWithBackoff;
var sso_web_oidc_base_1 = require("@mtfe/sso-web-oidc-base");
var random_1 = require("./random");
var net_1 = require("net");
var sleep = function (ms) {
    return new Promise(function (resolve) {
        setTimeout(resolve, ms);
    });
};
exports.sleep = sleep;
/**
 * 检查错误是否为ELOCKED类型
 * @param error 任意类型的错误
 * @returns 是否为ELOCKED错误
 */
function isELockedError(error) {
    if (error && typeof error === 'object' && 'code' in error) {
        return error.code === 'ELOCKED';
    }
    return false;
}
var builtError = function (error) {
    var errorMsg = '';
    if ((0, sso_web_oidc_base_1.tsSafeCheckKey)(error, 'msg')) {
        errorMsg = error.msg;
    }
    else if ((0, sso_web_oidc_base_1.tsSafeCheckKey)(error, 'message')) {
        errorMsg = error.message;
    }
    if (error && (0, sso_web_oidc_base_1.tsSafeCheckKey)(error, 'code'))
        return error;
    return (0, sso_web_oidc_base_1.buildResult)(sso_web_oidc_base_1.SSOErrorCode.UNKNOWN_ERROR, null, errorMsg);
};
exports.builtError = builtError;
/**
 * 检查指定端口是否可用
 * @param port 要检查的端口号
 * @returns Promise<boolean> 端口是否可用
 */
function checkPortAvailable(port) {
    return new Promise(function (resolve) {
        var server = (0, net_1.createServer)();
        server.listen(port, function () {
            server.close(function () {
                resolve(true);
            });
        });
        server.on('error', function () {
            resolve(false);
        });
    });
}
/**
 * 查找第一个可用端口
 * @param ports 端口数组，按优先级排序
 * @returns Promise<number | null> 第一个可用的端口号，如果都不可用则返回null
 */
function findFirstAvailablePort(ports) {
    return __awaiter(this, void 0, void 0, function () {
        var validPorts, _i, validPorts_1, port, isAvailable, _a;
        return __generator(this, function (_b) {
            switch (_b.label) {
                case 0:
                    // 参数验证
                    if (!Array.isArray(ports) || ports.length === 0) {
                        return [2 /*return*/, null];
                    }
                    validPorts = ports.filter(function (port) { return typeof port === 'number' && Number.isInteger(port) && port >= 1 && port <= 65535; });
                    if (validPorts.length === 0) {
                        return [2 /*return*/, null];
                    }
                    _i = 0, validPorts_1 = validPorts;
                    _b.label = 1;
                case 1:
                    if (!(_i < validPorts_1.length)) return [3 /*break*/, 6];
                    port = validPorts_1[_i];
                    _b.label = 2;
                case 2:
                    _b.trys.push([2, 4, , 5]);
                    return [4 /*yield*/, checkPortAvailable(port)];
                case 3:
                    isAvailable = _b.sent();
                    if (isAvailable) {
                        return [2 /*return*/, port];
                    }
                    return [3 /*break*/, 5];
                case 4:
                    _a = _b.sent();
                    // 忽略任何错误，继续检查下一个端口
                    return [3 /*break*/, 5];
                case 5:
                    _i++;
                    return [3 /*break*/, 1];
                case 6: return [2 /*return*/, null];
            }
        });
    });
}
/**
 * 重试函数，支持根据错误类型进行不同次数的重试
 * @param promiseFunction 要执行的Promise函数
 * @param logger 日志记录器
 * @param tag 日志标记，用于标识重试的操作
 * @returns Promise的返回值
 */
function retryWithBackoff(promiseFunction, logger, tag) {
    return __awaiter(this, void 0, void 0, function () {
        var ELOCKED_MAX_RETRIES, OTHER_ERROR_MAX_RETRIES, INITIAL_MIN_DELAY, INITIAL_MAX_DELAY, MAX_DELAY, elockedAttempts, otherErrorAttempts, currentMinDelay, currentMaxDelay, initialDelay, error_1, isElocked, errorMsg, delay;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    ELOCKED_MAX_RETRIES = 30;
                    OTHER_ERROR_MAX_RETRIES = 5;
                    INITIAL_MIN_DELAY = 100;
                    INITIAL_MAX_DELAY = 200;
                    MAX_DELAY = 5000;
                    elockedAttempts = 0;
                    otherErrorAttempts = 0;
                    currentMinDelay = INITIAL_MIN_DELAY;
                    currentMaxDelay = INITIAL_MAX_DELAY;
                    initialDelay = (0, random_1.getRandom)(INITIAL_MIN_DELAY, INITIAL_MAX_DELAY);
                    return [4 /*yield*/, (0, exports.sleep)(initialDelay)];
                case 1:
                    _a.sent();
                    _a.label = 2;
                case 2:
                    if (!true) return [3 /*break*/, 8];
                    _a.label = 3;
                case 3:
                    _a.trys.push([3, 5, , 7]);
                    return [4 /*yield*/, promiseFunction()];
                case 4: return [2 /*return*/, _a.sent()];
                case 5:
                    error_1 = _a.sent();
                    isElocked = isELockedError(error_1);
                    errorMsg = error_1 && typeof error_1 === 'object'
                        ? error_1.message || error_1.msg
                        : String(error_1);
                    if (isElocked) {
                        elockedAttempts++;
                        if (elockedAttempts > ELOCKED_MAX_RETRIES) {
                            logger.error(tag, "fail - ELOCKED error exceeded max retries (".concat(ELOCKED_MAX_RETRIES, "): ").concat(errorMsg));
                            throw (0, exports.builtError)(error_1);
                        }
                        logger.debug(tag, "ELOCKED error occurred, retrying (".concat(elockedAttempts, "/").concat(ELOCKED_MAX_RETRIES, "): ").concat(errorMsg));
                    }
                    else if (error_1 &&
                        (0, sso_web_oidc_base_1.tsSafeCheckKey)(error_1, 'code') &&
                        [
                            sso_web_oidc_base_1.SSOErrorCode.CREATE_SERVER_ERROR,
                            sso_web_oidc_base_1.SSOErrorCode.NO_AVALIABLE_PORT,
                            sso_web_oidc_base_1.SSOErrorCode.RETRY_TOO_MANY_TIME,
                        ].includes(error_1.code)) {
                        otherErrorAttempts += 1;
                        if (otherErrorAttempts > OTHER_ERROR_MAX_RETRIES) {
                            logger.error(tag, "fail - Error exceeded max retries (".concat(OTHER_ERROR_MAX_RETRIES, "): ").concat(error_1.code, " ").concat(errorMsg));
                            throw (0, exports.builtError)(error_1);
                        }
                        logger.debug(tag, "Error occurred, retrying (".concat(otherErrorAttempts, "/").concat(OTHER_ERROR_MAX_RETRIES, "):  ").concat(error_1.code, " ").concat(errorMsg));
                    }
                    else {
                        throw error_1;
                    }
                    delay = (0, random_1.getRandom)(currentMinDelay, currentMaxDelay);
                    logger.debug(tag, "Waiting ".concat(delay, "ms before retry..."));
                    return [4 /*yield*/, (0, exports.sleep)(delay)];
                case 6:
                    _a.sent();
                    // 为下次重试更新延迟时间（指数退避）
                    currentMinDelay = Math.min(currentMinDelay * 2, MAX_DELAY);
                    currentMaxDelay = Math.min(currentMaxDelay * 2, MAX_DELAY);
                    return [3 /*break*/, 7];
                case 7: return [3 /*break*/, 2];
                case 8: return [2 /*return*/];
            }
        });
    });
}
